---
title:  "CHIKV 28dpi scRNA-seq"
author: "Ryan Sheridan"
date:   "`r Sys.Date()`"
output: 
  html_document:
    toc:       true
    toc_float: true
    toc_depth: 2
    theme:     cosmo
    highlight: kate
params:
  
  unenr_dir:    "results/2021-11-05"      # Directory for unenriched data
  enr_dir:      "results/2022-02-03"      # Directory for CHIKV-enriched data
  template_dir: "src"                     # Directory for Rmd templates
  ref_dir:      "ref"
  table_dir:    "results/tables/2024-03-29"          # Directory to write tables
  geo_dir:      "results/geo"             # Directory for GEO submission
  overwrite:    false                     # Should tables be overwritten if they already exist?
  so_dir:       "~/Dropbox/Ryan/Projects/morrison-chronic-chikv/results/sobjs/2024-01-22_4" # Directory for Seurat objects
  metrics:     "count_metrics.csv"       # Cell Ranger metrics
  gene_min:    250                       # Min number of detected genes per cell
  gene_max:    5500                      # Max number of detected genes per cell
  mito_max:    30                        # Max percentage mito reads per cells
  rm_doublets: true                      # Remove doublets
  type_res:    5                         # Resolution for annotating cell types
  mac_res:     2                         # Resolution for annotating macrophages
  t_res:       1                         # Resolution for annotating T cells
  fib_res:     5                         # Resolution for annotating fibroblasts cells
  tm:          "28dpi"                   # Label for time point
  n_threads:   2
  rslns:
    value:
      type: 5
      mac:  2
      t:    1
      fib:  5
  samples:
    - "U1"
    - "U2"
    - "U3"
    - "C1"
    - "C2"
    - "C3"
  treatment_levels: ["mock", "CHIKV"]  # order as control, treatment, order affects how marker genes are identified
  xen_dir:          "~/Dropbox/Ryan/Projects/morrison-chronic-chikv/results/xenium"  # Xenium objects
  xen_res_dir:      "results/xenium/20241212__000859__Morrison_Run1"                 # Directory with Xenium data
  xen_regex:        ["output-XETG00230__00536[0-9]{2}__TR-2-24[A-Z]_"]               # Regex to pull sample dirs for Xenium slides
  xen_sample_regex: ["TR-2-24[A-Z]+"]
  xen_samples: 
    value:
      TR-2-24A: "A"
      TR-2-24B: "B"
      TR-2-24C: "C"
      TR-2-24D: "D"
      TR-2-24E: "E"
      TR-2-24F: "F"
  xen_chikv_targets:
    value:
      CHIKV_5:     "CHIKV-5-(0-7566)"
      CHIKV_sgRNA: "CHIKV-sgRNA-(7566-12036)"
      CHIKV_neg:   "CHIKV-negative-(0-12036)"
  xen_custom_targets:
    - "Ag-tag-A-(ERCC2)"
    - "Ag-tag-B-(ERCC3)"
    - "Ag-tag-C-(ERCC4)"
    - "T2-Sensitive-Ag-tag-(ERCC9)"
    - "T2-Resistant-Ag-tag-(ERCC12)"
    - "DNA-only-T2-Control-Ag-tag-(ERCC16)"
    - "GeoMx-Barcode01-ERCC00142"
    - "GeoMx-Barcode02-ERCC00144"
    - "CHIKV-5-(0-7566)"
    - "CHIKV-sgRNA-(7566-12036)"
    - "CHIKV-negative-(0-12036)"
    - "LCMV-(Armstrong)-Segment"
    - "Listeria-monocytogenes-10403S-gyrase-B-CP002002-1"
    - "Vaccinia-(Western-Reserve)-J6R"
    - "mRNA-vaccine-(3-UTR)"
    - "mRNA-vaccine-(ovalbumin)"
    - "mRNA-vaccine-(FLAG-3x)"
    - "mRNA-vaccine-(eGFP)"
editor_options: 
  chunk_output_type: console
---

<br>

```{r "setup", include = FALSE}
# Default chunk options
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo    = FALSE,
  dpi     = 300
)

options(future.globals.maxSize = 5.5 * 1024^3)

knitr::knit(here::here("src/setup.Rmd"), output = "")

# Columns to exclude when saving DE tables
exclude_clmns <- c(
  "max_pval", "minimump_p_val", "avg_fc",
  "all_same_direction", "n_same_direction",
  "significant"
)
```

```{r "format objects"}
# Calculate median fraction of CHIKV+ cells for clusters
so_mac <- so_mac %>%
  mutate_meta(~ {
    .x %>%
      group_by(!!sym(final_clsts$mac), treatment) %>%
      mutate(
        frac_chikv_cells = sum(chikv_grp == chikv_grps[2]) / n()
      ) %>%
      ungroup()
  })

# Set cutoff based on median fraction of CHIKV+ cells
# get the median for cells, not clusters, so low/high groups are similar size
frac_lim <- so_mac@meta.data %>%
  dplyr::filter(treatment == treats[2]) %>%
  pull(frac_chikv_cells) %>%
  median()
  
so_mac <- so_mac %>%
  mutate_meta(
    mutate,
    chikv_clst_grp = ifelse(
      frac_chikv_cells >= frac_lim,
      chikv_grps[2],
      chikv_grps[1]
    ),
    chikv_clst_grp = fct_relevel(chikv_clst_grp, chikv_grps)
  )
```

```{r "chikv markers"}
# Identify mock vs CHIKV DEGs for all cells
# significant genes have adjusted p-value < 0.05 and
# abs log2 fold change > 0.25 for all reps
file_nm <- here(dirs$table_dir, "chikv_markers")

Idents(so) <- so$treatment

chikv_degs <- so %>%
  find_conserved_markers(
    ident_1    = treats[2],
    ident_2    = treats[1],
    rep_var    = "rep",
    group_var  = NULL,
    
    p_max        = 0.05,
    fc_min       = 0.25,
    p_adj_method = "BH",
    n_reps       = 3,
    
    filter_sig = FALSE,
    file_path  = str_c(file_nm, ".tsv.gz"),
    n_threads  = 1
  )

# Save significant markers as xlsx file
info <- "
  Markers were identified for mock- vs CHIKV-infected mice for all cells.
  Genes were filtered for those with an absolute log2 fold change >0.25 for all
  replicates and a maximum adjusted p-value <0.05 (Benjamini-Hochberg).
  Mitochondrial genes were removed.
"

chikv_sig <- chikv_degs %>%
  filter(significant & !grepl("^mt-", gene)) %>%
  arrange(max_p_adj)

chikv_sig %>%
  dplyr::select(-any_of(exclude_clmns)) %>%
  write_xlsx(
    info      = info,
    tab_var   = "direction",
    file_path = str_c(file_nm, ".xlsx")
  )
```

```{r "chikv gsea"}
# all mock vs CHIKV-infected DEGs
# GSEA analysis using DOSE database
# need to fetch entrez gene IDs for human homologs
# exclude genes that have multiple human or mouse IDs
gsea_degs <- chikv_degs %>%
  inner_join(hlogs, by = c(gene = "gene_mm")) %>%
  distinct(gene, gene_hs, entrez_hs, min_abs_fc) %>%
  
  filter(if_all(
    -min_abs_fc,
    ~ !duplicated(.x) & !duplicated(.x, fromLast = TRUE)
  )) %>%
  arrange(desc(min_abs_fc))

# Run GSEA
gsea_file <- here(dirs$table_dir, "chikv_gsea.qs")

if (!file.exists(gsea_file)) {
  set.seed(42)
  
  gsea_res <- set_names(gsea_degs$min_abs_fc, gsea_degs$entrez_hs)
  
  chikv_gsea <- gsea_res %>%
    gseDO(eps = 1e-100) %>%
    setReadable("org.Hs.eg.db", "ENTREZID")
  
  chikv_gsea %>%
    qsave(gsea_file)
}

chikv_gsea <- qread(gsea_file)
```

```{r "cell type chikv markers"}
# Identify mock vs CHIKV DEGs for cell types
# significant genes have adjusted p-value < 0.05 and
# abs log2 fold change > 0.25 for all reps
file_nm <- here(dirs$table_dir, "celltype_chikv_markers")

Idents(so) <- so$treatment

type_chikv_degs <- so %>%
  find_conserved_markers(
    ident_1    = treats[2],
    ident_2    = treats[1],
    rep_var    = "rep",
    group_var  = "cell_type",
    
    p_max        = 0.05,
    fc_min       = 0.25,
    p_adj_method = "BH",
    n_reps       = 3,
    
    filter_sig = FALSE,
    file_path  = str_c(file_nm, ".tsv.gz"),
    n_threads  = 1
  )

# Save significant markers as xlsx file
info <- "
  Markers were identified for mock- vs CHIKV-infected mice for each cell type.
  Genes were filtered for those with an absolute log2 fold change >0.25 for all
  replicates and a maximum adjusted p-value <0.05 (Benjamini-Hochberg).
  Mitochondrial genes were removed.
"

type_chikv_sig <- type_chikv_degs %>%
  filter(significant & !grepl("^mt-", gene)) %>%
  arrange(max_p_adj)

type_chikv_sig %>%
  dplyr::select(-any_of(exclude_clmns)) %>%
  write_xlsx(
    info      = info,
    tab_var   = c("cell_type", "direction"),
    file_path = str_c(file_nm, ".xlsx")
  )
```

```{r "mac type markers"}
# Macrophage subset markers
# using all T cells identify markers for each subset
file_nm <- here(dirs$table_dir, "mac_type_markers")

Idents(so_mac) <- so_mac$mac_type

mac_typs <- unique(so_mac$mac_type)

mac_type_degs <- so_mac %>%
  find_conserved_markers(
    ident_1    = mac_typs,
    rep_var    = "rep",
    
    p_max        = 0.05,
    fc_min       = 0.25,
    p_adj_method = "BH",
    n_reps       = 3,
    
    filter_sig = FALSE,
    file_path  = str_c(file_nm, ".tsv.gz"),
    n_threads  = params$n_threads
  )

# Save significant markers as xlsx file
info <- "
  Markers were identified for each macrophage subset.
  Genes were filtered for those with an absolute log2 fold change >0.25 for all
  replicates and a maximum adjusted p-value <0.05 (Benjamini-Hochberg).
"

mac_type_sig <- mac_type_degs %>%
  filter(significant & direction == "up") %>%
  arrange(max_p_adj)

mac_type_sig %>%
  dplyr::select(-any_of(exclude_clmns)) %>%
  write_xlsx(
    info      = info,
    tab_var   = "ident_1",
    file_path = str_c(file_nm, ".xlsx")
  )
```

```{r "mac chikv markers"}
# all macroaphges mock vs CHIKV
# for GSEA include all genes regardless of significance
# significant genes have adjusted p-value < 0.05 and
# abs log2 fold change > 0.25 for all reps
file_nm <- here(dirs$table_dir, "mac_chikv_markers")

Idents(so_mac) <- so_mac$treatment

mac_chikv_degs <- so_mac %>%
  find_conserved_markers(
    ident_1    = treats[2],
    ident_2    = treats[1],
    rep_var    = "rep",
    
    p_max        = 0.05,
    fc_min       = 0.25,
    p_adj_method = "BH",
    n_reps       = 3,
    
    filter_sig = FALSE,
    file_path  = str_c(file_nm, ".tsv.gz"),
    n_threads  = params$n_threads
  )

# Save significant markers as xlsx file
info <- "
  Markers were identified for mock- vs CHIKV-infected mice for all macrophages.
  Genes were filtered for those with an absolute log2 fold change >0.25 for all
  replicates and a maximum adjusted p-value <0.05 (Benjamini-Hochberg).
  Mitochondrial genes were removed.
"

mac_chikv_sig <- mac_chikv_degs %>%
  filter(significant & !grepl("^mt-", gene)) %>%
  arrange(max_p_adj)

mac_chikv_sig %>%
  dplyr::select(-any_of(exclude_clmns)) %>%
  write_xlsx(
    info      = info,
    tab_var   = c("ident_1", "direction"),
    file_path = str_c(file_nm, ".xlsx")
  )
```

```{r "mac chikv gsea"}
# all macrophages mock vs CHIKV-infected DEGs
# GSEA analysis using DOSE database
# need to fetch entrez gene IDs for human homologs
# exclude genes that have multiple human or mouse IDs
gsea_degs <- mac_chikv_degs %>%
  inner_join(hlogs, by = c(gene = "gene_mm")) %>%
  distinct(gene, gene_hs, entrez_hs, min_abs_fc) %>%
  
  filter(if_all(
    -min_abs_fc,
    ~ !duplicated(.x) & !duplicated(.x, fromLast = TRUE)
  )) %>%
  arrange(desc(min_abs_fc))

gsea_file <- here(dirs$table_dir, "mac_chikv_gsea.qs")

if (!file.exists(gsea_file)) {
  set.seed(42)
  
  gsea_res <- set_names(gsea_degs$min_abs_fc, gsea_degs$entrez_hs)
  
  chikv_gsea <- gsea_res %>%
    gseDO(eps = 1e-100) %>%
    setReadable("org.Hs.eg.db", "ENTREZID")
  
  chikv_gsea %>%
    qsave(gsea_file)
}

mac_chikv_gsea <- qread(gsea_file)
```

```{r "mac type chikv markers"}
# Identify mock vs CHIKV DEGs for macroaphges subsets
# significant genes have adjusted p-value < 0.05 and
# abs log2 fold change > 0.25 for all reps
file_nm <- here(dirs$table_dir, "mac_type_chikv_markers")

Idents(so_mac) <- so_mac$treatment

mac_type_chikv_degs <- so_mac %>%
  find_conserved_markers(
    ident_1    = treats[2],
    ident_2    = treats[1],
    rep_var    = "rep",
    group_var  = "mac_type",
    
    p_max        = 0.05,
    fc_min       = 0.25,
    p_adj_method = "BH",
    n_reps       = 3,
    
    filter_sig = FALSE,
    file_path  = str_c(file_nm, ".tsv.gz"),
    n_threads  = params$n_threads
  )

# Save significant markers as xlsx file
info <- "
  Markers were identified for mock- vs CHIKV-infected mice for each macrophage
  subset.
  Genes were filtered for those with an absolute log2 fold change >0.25 for all
  replicates and a maximum adjusted p-value <0.05 (Benjamini-Hochberg).
  Mitochondrial genes were removed.
"

mac_type_chikv_sig <- mac_type_chikv_degs %>%
  filter(significant & !grepl("^mt-", gene)) %>%
  arrange(max_p_adj)

mac_type_chikv_sig %>%
  dplyr::select(-any_of(exclude_clmns)) %>%
  write_xlsx(
    info      = info,
    tab_var   = c("mac_type", "direction"),
    file_path = str_c(file_nm, ".xlsx")
  )
```

```{r "t type markers"}
# T cell subset markers
# using all T cells identify markers for each subset
file_nm <- here(dirs$table_dir, "t_type_markers")

Idents(so_t) <- so_t$t_type

t_typs <- unique(so_t$t_type)

t_type_degs <- so_t %>%
  find_conserved_markers(
    ident_1    = t_typs,
    rep_var    = "rep",
    p_max      = 0.05,
    fc_min     = 0.25,
    
    p_adj_method = "BH",
    
    n_reps     = 3,
    filter_sig = FALSE,
    file_path  = str_c(file_nm, ".tsv.gz"),
    n_threads  = params$n_threads
  )

# Save significant markers as xlsx file
info <- "
  Markers were identified for each T cell subset.
  Genes were filtered for those with an absolute log2 fold change >0.25 for all
  replicates and a maximum adjusted p-value <0.05 (Benjamini-Hochberg).
"

t_type_sig <- t_type_degs %>%
  filter(significant & direction == "up") %>%
  arrange(max_p_adj)

t_type_sig %>%
  dplyr::select(-any_of(exclude_clmns)) %>%
  write_xlsx(
    info      = info,
    tab_var   = "ident_1",
    file_path = str_c(file_nm, ".xlsx")
  )
```

```{r "t type go"}
# GO terms
# terms > 10 and < 500 genes
go_file <- here(dirs$table_dir, "t_type_go.tsv")

if (!file.exists(go_file)) {
  expr_gns <- so_t@assays$RNA@counts %>%
    rowSums() %>%
    sort(decreasing = TRUE) %>%
    head(5000)
  
  t_type_go <- t_type_sig %>%
    split(.$ident_1) %>%
    map(pull, gene) %>%
    gost(
      organism     = "mmusculus",
      custom_bg    = names(expr_gns),
      domain_scope = "custom",
      evcodes      = TRUE
    ) %>%
    .$result %>%
    dplyr::select(-c(evidence_codes, parents)) %>%
    filter(p_value < 0.05, term_size > 10, term_size < 500) %>%
    arrange(query, p_value)
  
  t_type_go %>%
    write_tsv(go_file)
}

t_type_go <- read_tsv(go_file)
```

```{r "t chikv markers"}
# CHIKV-infected DEGs for all T cells grouped together
file_nm <- here(dirs$table_dir, "t_chikv_markers")

Idents(so_t) <- so_t$treatment

t_degs_1 <- so_t %>%
  find_conserved_markers(
    ident_1    = treats[2],
    ident_2    = treats[1],
    rep_var    = "rep",
    
    p_max        = 0.05,
    fc_min       = 0.25,
    p_adj_method = "BH",
    n_reps       = 3,
    
    filter_sig = FALSE,
    file_path  = str_c(file_nm, ".tsv.gz"),
    n_threads  = params$n_threads
  )

# Save significant markers as xlsx file
info <- "
  Markers were identified for mock- vs CHIKV-infected mice for all T cells.
  Genes were filtered for those with an absolute log2 fold change >0.25 for all
  replicates and a maximum adjusted p-value <0.05 (Benjamini-Hochberg).
  Mitochondrial genes were removed.
"

t_sig_1 <- t_degs_1 %>%
  filter(significant & !grepl("^mt-", gene)) %>%
  arrange(max_p_adj)

t_sig_1 %>%
  dplyr::select(-any_of(exclude_clmns)) %>%
  write_xlsx(
    info      = info,
    tab_var   = c("ident_1", "direction"),
    file_path = str_c(file_nm, ".xlsx")
  )
```

```{r "t type chikv markers"}
# CHIKV-infected DEGs for each subset
file_nm <- here(dirs$table_dir, "t_type_chikv_markers")

Idents(so_t) <- so_t$treatment

if (!file.exists(str_c(file_nm, ".tsv.gz"))) {
  typs <- unique(so_t$t_type)
  
  t_degs_3 <- typs %>%
    map_dfr(~ {
      so_t %>%
        subset(t_type == .x) %>%
        FindMarkers(
          ident.1         = treats[2],
          ident.2         = treats[1],
          logfc.threshold = 0.25
        ) %>%
        as_tibble(rownames = "gene") %>%
        mutate(
          t_type    = .x,
          p_val_adj = p.adjust(p_val, method = "BH", n = nrow(so_t)),
          direction = ifelse(avg_log2FC > 0, "up", "down")
        ) %>%
        arrange(desc(avg_log2FC))
    })
  
  t_degs_3 %>%
    write_tsv(str_c(file_nm, ".tsv.gz"))
}

# Save significant markers as xlsx file
info <- "
  Upregulated genes were identified for mock- vs CHIKV-infected mice for each
  T subset.
  Due to the low number of T cells, all replicates were grouped together.
  Genes were filtered for those with an absolute log2 fold change >0.25,
  adjusted p-value <0.05 (Benjamini-Hochberg), expressed in >30% of cells from
  CHIKV-infected mice, and expressed in <70% of cells from mock-infected mice.
  Mitochondrial genes were removed.
"

t_degs_3 <- read_tsv(str_c(file_nm, ".tsv.gz"))

t_sig_3 <- t_degs_3 %>%
  filter(
    p_val_adj       < 0.05,
    abs(avg_log2FC) > 0.25,
    pct.1           > 0.3,
    pct.2           < 0.7,
    !grepl("^mt-", gene)
  ) %>%
  arrange(p_val_adj)

t_sig_3 %>%
  dplyr::select(-any_of(exclude_clmns)) %>%
  write_xlsx(
    info      = info,
    tab_var   = c("t_type", "direction"),
    file_path = str_c(file_nm, ".xlsx")
  )

# Only use upregulated genes for plots
t_sig_3 <- t_sig_3 %>%
  filter(direction == "up")
```

```{r "t CD4/8 chikv markers"}
# CHIKV-infected DEGs for CD4/8 T cells grouped together
# do not use replicates since we have so few T cells
file_nm <- here(dirs$table_dir, "t_type_chikv_markers-2")

Idents(so_t) <- so_t$treatment

if (!file.exists(str_c(file_nm, ".tsv.gz"))) {
  t_degs_2 <- c("CD4", "CD8") %>%
    map_dfr(~ {
      so_t %>%
        subset(cd4_8 == .x) %>%
        FindMarkers(
          ident.1         = treats[2],
          ident.2         = treats[1],
          logfc.threshold = 0.25
        ) %>%
        as_tibble(rownames = "gene") %>%
        mutate(
          cd4_8     = .x,
          p_val_adj = p.adjust(p_val, method = "BH", n = nrow(so_t)),
          direction = ifelse(avg_log2FC > 0, "up", "down")
        ) %>%
        arrange(desc(avg_log2FC))
    })
  
  t_degs_2 %>%
    write_tsv(str_c(file_nm, ".tsv.gz"))
}

# Save significant markers as xlsx file
info <- "
  Upregulated genes were identified for mock- vs CHIKV-infected mice for all
  CD4 and CD8 T cells.
  Due to the low number of T cells, all replicates were grouped together.
  Genes were filtered for those with an absolute log2 fold change >0.25,
  adjusted p-value <0.05 (Benjamini-Hochberg), expressed in >30% of cells from
  CHIKV-infected mice, and expressed in <70% of cells from mock-infected mice.
  Mitochondrial genes were removed.
"

t_degs_2 <- read_tsv(str_c(file_nm, ".tsv.gz"))

t_sig_2 <- t_degs_2 %>%
  filter(
    p_val_adj       < 0.05,
    abs(avg_log2FC) > 0.25,
    pct.1           > 0.3,
    pct.2           < 0.7,
    !grepl("^mt-", gene)
  ) %>%
  arrange(p_val_adj)

t_sig_2 %>%
  dplyr::select(-any_of(exclude_clmns)) %>%
  write_xlsx(
    info      = info,
    tab_var   = c("cd4_8", "direction"),
    file_path = str_c(file_nm, ".xlsx")
  )

t_sig_2 <- t_sig_2 %>%
  filter(direction == "up")
```

# Page 1

## 1C-D

UMAP projections show sample groups (left), cell type annotations (middle),
and CHIKV+ cells (right).

```{r "cell type UMAPs", fig.height = 5, fig.width = 18}
# UMAP showing samples
u1 <- so %>%
  plot_scatter(
    "sample",
    "UMAP_1", "UMAP_2",
    size = 0.001,
    plot_colors = sam_cols,
    plot_lvls   = rev(sam_lvls)
  ) +
  umap_theme +
  theme(
    panel.border = element_rect(color = ln_col, linewidth = ln_pt),
    legend.title = element_blank()
  )

# UMAP showing cell types
u2 <- so %>%
  plot_scatter(
    "cell_type",
    "UMAP_1", "UMAP_2",
    size = 0.001,
    plot_colors = type_cols
  ) +
  umap_theme +
  theme(
    panel.border = element_rect(color = ln_col, linewidth = ln_pt),
    legend.title = element_blank()
  )
  
# UMAP showing CHIKV+ cells
u_labs <- so@meta.data %>%
  filter(chikv_grp == chikv_grps[2]) %>%
  group_by(cell_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  arrange(desc(n)) %>%
  mutate(
    n = str_c(cell_type, "\nn = ", n)
  )

u_labs <- set_names(u_labs$n, u_labs$cell_type)

u3 <- so@meta.data %>%
  mutate(cell_type = fct_relevel(cell_type, names(u_labs))) %>%
  ggplot(aes(UMAP_1, UMAP_2)) +
  geom_point_trace(
    trace_position = "bottom",
    fill = "white",
    stroke = 0.7,
    size = 0.5
  ) +
  geom_point(
    aes(color = cell_type),
    data = ~ filter(.x, chikv_grp == chikv_grps[2]),
    size = 1
  ) +
  guides(color = guide_legend(override.aes = list(size = 3))) +
  scale_color_manual(values = type_cols, labels = u_labs) +
  umap_theme +
  theme(
    panel.border = element_rect(color = ln_col, linewidth = ln_pt),
    legend.title = element_blank()
  )

# Create final figure
plot_grid(
  u1, u2, u3,
  nrow = 1,
  align = "h",
  axis  = "tb"
)
```

<br>

The fraction of cells identified for each cell type is shown below.

```{r "cell type bargraphs", fig.width = 5, fig.height = 4, out.width = "50%"}
so@meta.data %>%
  mutate(
    cell_type = fct_infreq(cell_type),
    treatment = fct_relevel(treatment, treats)
  ) %>%
  ggplot(aes(rep, fill = cell_type)) +
  geom_bar(position = "fill") +
  facet_wrap(~ treatment) +
  scale_fill_manual(values = type_cols) +
  scale_y_continuous(breaks = c(0, 0.5, 1)) +
  labs(y = "fraction of cells") +
  djvdj_theme() +
  theme(
    aspect.ratio = 2.5,
    strip.text   = element_text(size = ttl_pt2),
    legend.title = element_blank(),
    legend.text  = element_text(size = txt_pt2),
    panel.border = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size = ttl_pt2),
    axis.text.x  = element_blank(),
    axis.text.y  = element_text(size = txt_pt2),
    axis.ticks.x = element_blank()
  )
```

<br>

## 1E

The number of differentially expressed genes identified for mock vs CHIKV is
shown below for each cell type.

```{r "DEGs bargraph", fig.width = 5, fig.height = 3, out.width = "%40"}
type_chikv_sig %>%
  mutate(
    cell_type = fct_infreq(cell_type),
    direction = fct_relevel(direction, c("up", "down"))
  ) %>%
  ggplot(aes(cell_type, fill = direction)) +
  geom_bar(color = "black", width = 0.75, linewidth = 0.35) +
  scale_fill_manual(values = c(up = sam_cols[[3]], down = sam_cols[[4]])) +
  labs(y = "# of DEGs") +
  base_theme +
  theme(
    aspect.ratio         = 0.6,
    legend.position      = c(0.5, 0.8),
    legend.title         = element_blank(),
    legend.direction     = "horizontal",
    legend.justification = "center",
    axis.line.x          = element_line(color = "black"),
    axis.line.y          = element_line(color = "black"),
    axis.ticks.x         = element_line(color = "black"),
    axis.ticks.y         = element_line(color = "black"),
    axis.title.x         = element_blank(),
    axis.text.x          = element_text(angle = 45, hjust = 1)
  )
```

<br>

## 1F

Bargraphs showing select GSEA disease ontology terms for genes upregulated when
comparing all mock vs CHIKV-infected macrophages.
This shows the top terms sorted by adjusted p value, with
"bone inflammation disease" added to the plot.
GSEA terms were filtered to remove terms that include >400 genes.

```{r "mac GSEA bargraph", fig.width = 7, fig.height = 4, out.width = "50%"}
# Bargraph showing top terms
# Filter to remove broad gene sets
select_terms <- c(
  "rheumatic disease",
  "rheumatoid arthritis",
  "bone inflammation disease"
)

n_terms <- 6

top_gsea <- mac_chikv_gsea@result %>%
  filter(setSize <= 400) %>%
  arrange(p.adjust) %>%
  filter(row_number() <= n_terms | Description %in% select_terms) %>%
  mutate(Description = fct_inorder(Description)) %>%
  arrange(desc(Description)) %>%  # reverse axis order so top shown first
  mutate(Description = fct_inorder(Description))
  
top_gsea %>%
  ggplot(aes(Description, -log10(p.adjust))) +
  geom_col(fill = "#6A51A3", alpha = 0.7) +
  labs(y = "-log10(p-value)") +
  scale_y_continuous(expand = expansion(c(0.02, 0.05))) +
  coord_flip() +
  djvdj_theme() +
  theme(
    aspect.ratio = 0.5,
    panel.border = element_blank(),
    axis.line.x  = element_line(size = 0.5, color = "grey85"),
    axis.ticks.y = element_blank(),
    axis.title.y = element_blank(),
    axis.text.y  = element_text(size = 20, hjust = 1, color = "black"),
    axis.text.x  = element_text(size = 16, hjust = 1, color = "black"),
    axis.title.x = element_text(size = 20)
  )
```

Enrichment is shown for select top GSEA terms

```{r "mac GSEA examples", fig.width = 10, fig.height = 5}
# Plot example terms
gsea_terms <- set_names(top_gsea$ID, top_gsea$Description)

gsea_terms %>%
  imap(~ {
    mac_chikv_gsea %>%
      gseaplot(
        .x,
        by = "runningScore",
        color.line = "#0072B2"
      ) +
      labs(subtitle = .y, x = "Gene rank") +
      theme_cowplot() +
      djvdj_theme() +
      theme(
        aspect.ratio = 0.65,
        plot.subtitle = element_text(face = "plain", size = 15),
        axis.title = element_text(size = 10)
      )
  }) %>%
  plot_grid(
    plotlist = .,
    nrow = 2
  )
```

---

<br>

<br>

# Page 2

## 2B

The fraction of CHIKV+ cells identified for each cell type is shown below.

```{r "CHIKV+ type bargraphs", fig.width = 5, fig.height = 4, out.width = "50%"}
dat <- so@meta.data %>%
  filter(chikv_grp == "CHIKV-high") %>%
  mutate(cell_type = fct_infreq(cell_type))

lgnd_labs <- dat %>%
  group_by(cell_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(lab = str_c(cell_type, " (n = ", label_comma()(n), ")"))

lgnd_labs <- set_names(lgnd_labs$lab, lgnd_labs$cell_type)

dat %>%
  ggplot(aes(rep, fill = cell_type)) +
  geom_bar(position = "fill") +
  scale_fill_manual(values = type_cols, labels = lgnd_labs) +
  scale_y_continuous(breaks = c(0, 0.5, 1)) +
  labs(y = "fraction of CHIKV+ cells") +
  djvdj_theme() +
  theme(
    aspect.ratio = 2.5,
    strip.text   = element_text(size = ttl_pt2),
    legend.title = element_blank(),
    legend.text  = element_text(size = txt_pt2),
    panel.border = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size = ttl_pt2),
    axis.text.x  = element_blank(),
    axis.text.y  = element_text(size = txt_pt2),
    axis.ticks.x = element_blank()
  )
```

<br>

The fraction of cells identified as CHIKV+ is shown below for each macrophage
subset from the CHIKV-infected samples.
Unassigned cells are not shown.

```{r "CHIKV+ mac bargraphs", fig.width = 5, fig.height = 3, out.width = "50%"}
# Format data for bargraphs
dat <- so_mac@meta.data %>%
  filter(treatment == "CHIKV" & mac_type != "unassigned") %>%
  
  group_by(mac_type) %>%
  summarize(
    n = n(),
    n_high = sum(chikv_grp == "CHIKV-high"),
    .groups = "drop"
  ) %>%
  mutate(
    frac_high = n_high / n,
    pct_high  = frac_high * 100,
    n_lab     = str_c(label_comma()(n_high), "/", label_comma()(n)),
    mac_type  = fct_relevel(mac_type, names(mac_x_labs))
  )

# Create bargraphs
dat %>%
  filter(mac_type != "unassigned") %>%
  
  ggplot(aes(mac_type, frac_high, fill = mac_type, color = mac_type)) +
  geom_col(position = position_dodge2(), alpha = 0.5) +
  geom_text(
    aes(label = n_lab),
    position = position_dodge2(width = 0.9),
    alpha = 1, color = "black",
    hjust = 0.5, vjust = -0.3,
    size = 8 / .pt
  ) +
  scale_fill_manual(values = mac_typ_cols_2) +
  scale_color_manual(values = mac_typ_cols_2) +
  scale_x_discrete(labels = mac_x_labs) +
  scale_y_continuous(labels = label_percent(), expand = expansion(c(0.05, 0.1))) +
  # scale_alpha_manual(values = rep(0.5, n_distinct(dat$rep))) +
  guides(alpha = "none") +
  labs(y = "% CHIKV+") +
  base_theme +
  theme(
    aspect.ratio = 1,
    legend.position = "none",
    axis.title.x = element_blank()
  )
```

---

<br>

<br>

# Page 4

## 4A-B

UMAP projections show macrophage subsets.

```{r "mac subset plots", fig.width = 11.25, fig.height = 5.6}
# Create mac subset UMAPs
create_subset_umap <- function(so_in, sub_clmn, clrs) {
  
  sub_labs <- get_nlab_fun(so_in, sub_clmn, l = " (", r = ")")
  
  res <- so_in %>%
    plot_scatter(
      sub_clmn,
      x = "hUMAP_1",
      y = "hUMAP_2",
      plot_lvls = names(clrs),
      size = 0.7
    ) +
    scale_color_manual(values = clrs, labels = sub_labs) +
    guides(color = guide_legend(
      ncol = 1, override.aes = list(size = 3), reverse = TRUE)
    ) +
    int_u_theme +
    theme(
      plot.title  = element_text(hjust = 0.5, size = 20),
      legend.text = element_text(size = 14)
    )
  
  res
}

# # Adjust mac_type labels
# so_mac <- so_mac %>%
#   mutate_meta(
#     mutate,
#     mac_type = replace_na(mac_type, "unassigned"),
#     mac_type = recode(mac_type, !!!mac_typ_labs)
#   )

# Create mac subset UMAPs
u <- treats %>%
  map(~ {
    so_mac %>%
      subset(treatment == .x) %>%
      create_subset_umap("mac_type", mac_typ_cols_2) +
      ggtitle(.x)
  })

# Create mac subset bargraphs
br <- so_mac@meta.data %>%
  mutate(
    treatment = fct_relevel(treatment, treats),
    rep = str_c("r", rep)
  ) %>%
  ggplot(aes(rep, fill = mac_type)) +
  geom_bar(position = "fill") +
  facet_wrap(~ treatment, scales = "free") +
  scale_fill_manual(values = mac_typ_cols_2) +
  scale_y_continuous(expand = expansion(0)) +
  scale_x_discrete(labels = (function(x) str_remove(x, str_c("-", params$tm)))) +
  base_theme +
  theme(
    aspect.ratio    = 3,
    legend.position = "none",
    strip.text      = element_text(size = 20),
    axis.title      = element_blank(),
    axis.text.y     = element_blank(),
    axis.line.x     = element_blank(),
    axis.line.y     = element_blank(),
    axis.ticks.x    = element_blank(),
    axis.ticks.y    = element_blank()
  )

# Create final figure
plot_grid(
  plotlist = append(u, list(br)),
  nrow  = 1,
  align = "vh",
  axis  = "trbl",
  rel_widths = c(1, 1, 0.7)
)

# Boxplots showing module score for assigned subsets
# so_mac@meta.data %>%
#   pivot_longer(all_of(feat_lvls)) %>%
#   ggplot(aes(mac_type, value, fill = mac_type)) +
#   geom_boxplot() +
#   facet_wrap(~ name, scales = "free_y") +
#   scale_fill_manual(values = mac_typ_cols) +
#   base_theme +
#   theme(
#     axis.text.x     = element_text(angle = 45, hjust = 1),
#     axis.title      = element_blank(),
#     legend.position = "none"
#   )

```

<br>

## 4C

UMAP projections show CHIKV+ cells (left) and CHIKV-high clusters
(middle, right) for CHIKV-infected samples.

```{r "CHIKV+ clusters", fig.width = 10, fig.height = 4.5}
# Set plot colors
chikv_clst_grp_clrs <- c(
  `CHIKV-high` = "#035B8F",
  `CHIKV-low`  = "#009E73"
)

# Create UMAP showing CHIKV+ cells
u_dat <- so_mac@meta.data %>%
  filter(treatment == "CHIKV")

u1 <- u_dat %>%
  ggplot(aes(hUMAP_1, hUMAP_2)) +
  
  geom_point_trace(
    trace_position = "bottom",
    fill = "white",
    size = 1.5
  ) +
  geom_point(
    aes(color = mac_type),
    data = ~ filter(.x, chikv_grp == chikv_grps[2]),
    size = 2
  ) +
  scale_color_manual(values = mac_typ_cols_2) +
  guides(color = guide_legend(nrow = 3)) +
  umap_theme +
  theme(
    legend.position = "bottom",
    legend.title    = element_blank(),
    legend.text     = element_text(size = 12),
    panel.border    = element_rect(color = ln_col, linewidth = ln_pt)
  )

# Create UMAP showing CHIKV-high clusters
u2 <- u_dat %>%
  plot_scatter(
    "chikv_clst_grp",
    "hUMAP_1", "hUMAP_2",
    size        = 1,
    plot_colors = chikv_clst_grp_clrs,
    n_label     = "legend"
  ) +
  umap_theme +
  theme(
    legend.position = "bottom",
    legend.title    = element_blank(),
    legend.text     = element_text(size = 12),
  )

# Bargraph showing CHIKV-low/high clusters
br <- u_dat %>%
  ggplot(aes(rep, fill = mac_type)) +
  geom_bar(position = "fill") +
  facet_wrap(~ chikv_clst_grp, nrow = 1) +
  labs(y = "fraction of cells") +
  scale_fill_manual(values = mac_typ_cols_2) +
  scale_y_continuous(breaks = c(0, 0.5, 1), expand = expansion(c(0, 0))) +
  base_theme +
  theme(
    aspect.ratio    = 2.5,
    legend.position = "none",
    strip.text      = element_text(size = 12),
    axis.title.x    = element_blank()
  )

# Create final figure
plot_grid(
  u1, u2, br,
  nrow  = 1,
  align = "h",
  axis  = "tb",
  rel_widths = c(1, 1, 0.7)
)
```

<br>

## 4D

Heatmaps show expression of the top 20 genes
upregulated in CHIKV- vs mock-infected samples for each macrophage subset.
Genes are arranged by the number of subsets upregulating the gene and
the maximum adjusted p-value for the replicates.
Black diamonds mark the top 20 upregulated genes for each subset,
grey diamonds mark other subsets where each gene is upregulated.

```{r "mac chikv type heatmaps", fig.width = 5.3, fig.height = 6, out.width = "75%"}
# Set genes to plot
# * exclude unassigned macrophages
# * filter expr_dat after selecting genes to ensure diamonds are included for
#   all cell types that significantly upregulate the gene
degs <- mac_type_chikv_sig %>%
  filter(direction == "up")

so_mac %>%
  create_deg_heatmap(
    type_clmn  = "mac_type",
    type_lvls  = names(mac_typ_cols),
    n_top      = 20,
    sam_clmn   = "sample",
    sam_lvls   = sam_lvls,
    p_data     = degs,
    
    p_clmn     = "max_p_adj",
    rev_sort   = FALSE,
    
    strip_labs = mac_strip_labs
  )
```

<br>

This is an alternate version showing the top 30 genes sorted by the maximum
fold change in expression for the replicates.
This modified version includes Il1b and Tnf.

```{r "mac chikv type heatmaps 2", fig.width = 5.5, fig.height = 9, out.width = "75%"}
# Set genes to plot
# * exclude unassigned macrophages
# * filter expr_dat after selecting genes to ensure diamonds are included for
#   all cell types that significantly upregulate the gene
degs <- mac_type_chikv_sig %>%
  filter(direction == "up")

so_mac %>%
  create_deg_heatmap(
    type_clmn  = "mac_type",
    type_lvls  = names(mac_typ_cols),
    n_top      = 30,
    sam_clmn   = "sample",
    sam_lvls   = sam_lvls,
    p_data     = degs,
    
    p_clmn     = "max_abs_fc",
    rev_sort   = TRUE,
    
    strip_labs = mac_strip_labs
  )
```

---

<br>

<br>

# Page 5

## 5A

Expression of select genes upregulated in CHIKV- vs mock-infected samples is
shown below for macrophage subsets.

```{r "mac chikv type boxplots", fig.width = 19, fig.height = 7.5}
# Genes to plot
gns <- c("Tnf", "Il1b", "Nlrp3", "H2-Eb1")

u_clrs    <- grp_cols
u_clrs[1] <- lighten("#009E73", 0.2)

names(u_clrs) <- treats

# Create plots for example genes
# * exclude unassigned macrophages
gn_boxes <- so_mac %>%
  subset(mac_type != "unassigned") %>%
  create_celltype_plots(
    gns     = gns,
    p_data  = mac_type_chikv_sig,
    p_clmn  = "max_p_adj",
    bx_labs = mac_x_labs
  )

ex_plts <- plot_grid(
  plotlist = gn_boxes,
  align = "v",
  axis  = "rl",
  nrow  = 2
)

ex_plts
```

<br>

## 5D

Heatmaps show expression of the top 10 marker genes for each macrophage subset.
Genes are sorted by minimum fold change for the replicates.

```{r "mac type heatmaps", fig.width = 5.2, fig.height = 6.3}
so_mac %>%
  create_marker_heatmap(
    type_clmn  = "mac_type",
    type_lvls  = names(mac_typ_cols),
    sam_clmn   = "sample",
    sam_lvls   = sam_lvls,
    p_data     = mac_type_sig,
    p_clmn     = "min_abs_fc",
    strip_labs = mac_strip_labs,
    n_gns      = 10
  )
```

---

<br>

<br>

# Page 6

## 6A

Ifng expression is shown below for each cell type for CHIKV-infected samples.
CD4 Teffs are plotted as a separate group.

```{r "cell type ifng", fig.width = 4, fig.height = 4.7, out.width = "65%"}
# Format Ifng data
ifng_dat <- so %>%
  mutate_meta(
    mutate,
    new_type = case_when(
      t_type == "CD4 Teff"                          ~ t_type,
      cell_type == "T cells" & t_type != "CD4 Teff" ~ "other T cells",
      TRUE                                          ~ cell_type
    )
  )

lvls <- ifng_dat %>%
  FetchData(c("cell_type", "new_type", "Ifng")) %>%
  group_by(cell_type, new_type) %>%
  summarize(
    med = median(Ifng),
    q3  = quantile(Ifng, 0.75),
    mx  = max(Ifng),
    .groups = "drop"
  ) %>%
  arrange(desc(med), desc(q3), desc(mx)) %>%
  mutate(
    rank      = row_number(),
    cell_type = fct_reorder(cell_type, rank, min)
  ) %>%
  arrange(cell_type) %>%
  mutate(new_type = fct_inorder(new_type))

# Plot Ifng expression
clrs <- type_cols
clrs[c("CD4 Teff", "other T cells")] <- type_cols["T cells"]

clrs <- clrs[as.character(unique(lvls$new_type))]

ifng_dat %>%
  subset(treatment == "CHIKV") %>%
  create_celltype_plots_2(
    gn       = "Ifng",
    x        = "UMAP_1",
    y        = "UMAP_2",
    bx_clmn  = "new_type",
    grp_clmn = "treatment",
    grp_lvls = treats,
    u_clrs   = c("#FFFFFF", "#035B8F"),
    bx_clrs  = t_clrs,
    rel_h    = c(1, 0.7),
    out_size = 0.1,
    pt_size  = 0.1
  )
```

<br>

## 6B

UMAP projections show T cell subsets.

```{r "t subset plots", fig.width = 9, fig.height = 5.6}
# Create mac subset UMAPs
create_subset_umap <- function(so_in, sub_clmn, clrs) {
  
  sub_labs <- get_nlab_fun(so_in, sub_clmn, l = " (", r = ")")
  
  res <- so_in %>%
    plot_scatter(
      sub_clmn,
      x = "hUMAP_1",
      y = "hUMAP_2",
      plot_lvls = names(clrs),
      size = 1.3
    ) +
    scale_color_manual(values = clrs, labels = sub_labs) +
    guides(color = guide_legend(
      ncol = 1, override.aes = list(size = 3), reverse = TRUE)
    ) +
    int_u_theme +
    theme(
      panel.border = element_rect(color = ln_col, linewidth = ln_pt),
      plot.title   = element_text(hjust = 0.5, size = 20),
      legend.text  = element_text(size = 14)
    )
  
  res
}

# Create subset UMAPs
u <- treats %>%
  map(~ {
    so_t %>%
      subset(treatment == .x) %>%
      create_subset_umap("t_type", t_clrs) +
      ggtitle(.x)
  })

# Create mac subset bargraphs
br <- so_t@meta.data %>%
  mutate(
    treatment = fct_relevel(treatment, treats),
    rep = str_c("r", rep)
  ) %>%
  ggplot(aes(rep, fill = t_type)) +
  geom_bar(position = "fill") +
  facet_wrap(~ treatment, scales = "free") +
  scale_fill_manual(values = t_clrs) +
  scale_y_continuous(expand = expansion(0)) +
  scale_x_discrete(labels = (function(x) str_remove(x, str_c("-", params$tm)))) +
  base_theme +
  theme(
    aspect.ratio    = 3,
    legend.position = "none",
    strip.text      = element_text(size = 20),
    axis.title      = element_blank(),
    axis.text.y     = element_blank(),
    axis.line.x     = element_blank(),
    axis.line.y     = element_blank(),
    axis.ticks.x    = element_blank(),
    axis.ticks.y    = element_blank()
  )

# Create final figure
plot_grid(
  plotlist = append(u, list(br)),
  nrow  = 1,
  align = "vh",
  axis  = "trbl",
  rel_widths = c(1, 1, 0.7)
)
```

<br>

## 6C

Heatmaps show expression of the top 15 genes
upregulated in CHIKV- vs mock-infected samples for each T cell subset.
Genes are arranged by the number of cell types upregulating the gene and
the maximum adjusted p-value for the replicates.
Black diamonds mark the top 15 upregulated genes for each subset,
grey diamonds mark other subsets where each gene is upregulated.

```{r "t chikv type heatmaps", fig.width = 6.5, fig.height = 6.75}
# Merge DEGs to also include unsbusetted CD8 and CD4 T cells
t_chikv_degs <- t_sig_2 %>%
  rename(t_type = cd4_8) %>%
  bind_rows(t_sig_3) %>%
  filter(direction == "up")

# Generate T object that also includes T subsets and unsubsetted T cells
so_t_degs <- so_t %>%
  mutate_meta(~ mutate(.x, t_type = cd4_8)) %>%
  merge(so_t)

t_deg_typs <- unique(so_t_degs$t_type)
t_deg_typs <- t_deg_typs[t_deg_typs %in% t_chikv_degs$t_type]

# Format T subset labels
t_strip_labs <- c(
  CD8 = "all~CD8",
  CD4 = "all~CD4",
  `CD8 naive` = "'CD8 T'['naive']",
  `CD8 Teff`  = "'CD8 T'['eff']",
  Tgd         = "'T'[gamma*delta]",
  `CD4 naive` = "'CD4 T'['naive']",
  `CD4 Teff`  = "'CD4 T'['eff']",
  Treg        = "'T'['reg']"
)

# Create heatmaps
# * create separate heatmap for each subset need to scale each subset separately
#   and thus need separate colorbar for each subset
t_heats <- t_deg_typs %>%
  imap(~ {
    plt <- so_t_degs %>%
      subset(t_type == .x) %>%
      create_deg_heatmap(
        type_clmn     = "t_type",
        type_lvls     = names(t_strip_labs),
        sam_clmn      = "sample",
        sam_lvls      = sam_lvls,
        p_data        = t_chikv_degs,
        p_clmn        = "p_val_adj",
        n_gns         = 15,
        inc_type      = t_deg_typs,
        strip_labs    = t_strip_labs
      )
    
    if (.y > 1) {
      plt <- plt +
        theme(
          axis.text.y  = element_blank(),
          axis.ticks.y = element_blank(),
          axis.line.y  = element_blank()
        )
    }
    
    plt
  })

# Pull legend to add separately
lgnd <- t_heats[[1]] +
  guides(fill = "none") +
  theme(legend.margin = margin(15, 15, 15, 15))

lgnd <- get_legend(lgnd)

# Create final figure
t_heats <- t_heats %>%
  wrap_plots(nrow = 1) &
  guides(
    color = "none",
    fill = guide_colorbar(
      ticks = FALSE,
      label.hjust = c(0, 1),
      label.vjust = 3
    )
  ) &
  theme(
    plot.margin          = margin(),
    legend.position      = "bottom",
    legend.key.height    = unit(3, "pt"),
    legend.key.width     = unit(11, "pt"),
    legend.justification = 0.5,
    legend.text          = element_text(hjust = c(0, 1), vjust = 3)
  )

wrap_plots(
  t_heats, lgnd,
  nrow = 1,
  widths = c(1, 0.2)
)
```

<br>

## 6D

Ifng expression is shown below for each T cell subset.

```{r "t ifng", fig.width = 5.75, fig.height = 3.75, out.width = "65%"}
so_t %>%
  create_celltype_plots_2(
    gn       = "Ifng",
    bx_clmn  = "t_type",
    grp_clmn = "treatment",
    grp_lvls = treats,
    u_clrs   = c("#FFFFFF", "#035B8F"),
    bx_clrs  = t_clrs,
    pt_size  = 1
  )
```

---

<br>

<br>

# Page 7

## 7A

T cell marker genes are shown below for mock- and CHIKV-infected samples.

```{r "t markers", fig.height = 20, fig.width = 20}
# Genes to plot
key_gns <- c(
  "Cd3e", "Cd4", "Cd8a",
  "Sell", "Ccr7", "Il7r",
  "Cd44", "Cd69", "Trdc",
  "Foxp3", "Tbx21", "Gzmb"
)

t_plts <- key_gns %>%
  map(~ {
    create_celltype_plots_2(
      so_in    = so_t,
      gn       = .x,
      bx_clmn  = "t_type",
      grp_clmn = "treatment",
      grp_lvls = treats,
      pt_size  = 1,
      bx_clrs  = t_clrs,
      ttl      = .x
    )
  })

# Create final figure
t_plts %>%
  plot_grid(
    plotlist = .,
    ncol     = 3,
    align    = "vh",
    axis     = "trbl"
  )
```

---

<br>

<br>

# Page 11

## 11A

Heatmaps show expression of the top 10 marker genes for each macrophage subset.
Genes are sorted by minimum fold change for the replicates.

```{r "t type heatmaps", fig.width = 7, fig.height = 9}
so_t %>%
  create_marker_heatmap(
    type_clmn  = "t_type",
    type_lvls  = names(t_strip_labs),
    sam_clmn   = "sample",
    sam_lvls   = sam_lvls,
    p_data     = t_type_sig,
    p_clmn     = "min_abs_fc",
    n_gns      = 10,
    strip_labs = t_strip_labs
  )
```

---

<br>

<br>

# Session info

```{r}
sessionInfo()
```
