---
title: "Chronic CHIKV Xenium analysis"
author: "Ryan Sheridan"
date: "`r Sys.Date()`"
output:
  html_document:
    toc:            true
    toc_float:      true
    toc_depth:      2
    df_print:       "paged"
    code_folding:   "hide"
    self_contained: true
params:
  template_dir: "src"                     # Directory for Rmd templates
  ref_dir:      "ref"
  table_dir:    "results/tables"          # Directory to write tables
  overwrite:    false                     # Should tables be overwritten if they already exist?
  so_dir:       "~/Dropbox/Ryan/Projects/morrison-chronic-chikv/results/sobjs/2024-01-22_4" # Directory for Seurat objects
  geo_dir:      "~/Projects/morrison-chronic-chikv/results/geo"
  treatment_levels: ["mock", "CHIKV"]  # order as control, treatment, order affects how marker genes are identified
  xen_obj_dir:      "~/Dropbox/Ryan/Projects/morrison-chronic-chikv/results/xenium"  # Directory to save Xenium objects
  xen_ref:          "~/Dropbox/Ryan/Projects/morrison-chronic-chikv/results/sobjs/2024-01-22_4/so.qs"
  xen_data_dir:     "results/xenium/20241212__000859__Morrison_Run1"                 # Directory with Xenium data
  xen_data_regex:
  xen_sample_regex: "TR-2-24[A-Z]+"                                                  # Regex to extract sample IDs from directories
  xen_sample_names:
    value:
      TR-2-24A: "mock-1"                                                             # New sample names for extracted sample IDs
      TR-2-24B: "mock-2"
      TR-2-24C: "mock-3"
      TR-2-24D: "CHIKV-1"
      TR-2-24E: "CHIKV-2"
      TR-2-24F: "CHIKV-3"
  xen_targets:      "ref/QX42XM_mMulti_98g_custom_sequences.csv"                     # Custom non-endogenous targets in panel
  xen_key_targets:
    value:
      CHIKV_sgRNA: "CHIKV-sgRNA-(7566-12036)"
      CHIKV_5:     "CHIKV-5-(0-7566)"                                                # Custom targets to quantify
      CHIKV_neg:   "CHIKV-negative-(0-12036)"
editor_options: 
  chunk_output_type: console
---

<br>

## Summary

* Cell types were annotated using CHIKV 28 dpi scRNA-seq data
* Macrophages harbor CHIKV RNA
* Signal was detected for CHIKV 3' (sgRNA), 5' (nsP genes), and negative strand probes
* The strongest CHIKV signal was detected for 3' probes

<u>To do</u>

- [x] &nbsp;Heatmap of all of the ankles showing spatial distribution of each
  cell type.
- [x] &nbsp;For CHIKV infection sections, label the total number of CHIKV RNA+
  cells for each individual ankle instead of all of them together.
  As well as graph the total number of CHIKV RNA+ cells for each cell subset
  instead of the % CHIKV RNA+ cells.
- [x] &nbsp;For that same section, make a cutoff of 20 cells so any cell types
  with fewer than 20 CHIKV+ cells would be excluded.
- [x] &nbsp;Change "low" to CHIKV RNA negative
- [x] &nbsp;Change the CHIKV RNA species bar graphs to only include cell types
  that are above the 0.5 confidence interval.
  So take out unassigned.
- [x] &nbsp;Change the subgenomic label in the axis to 3' and take out the mock
  samples.
- [ ] &nbsp;Double check that cells labeled “DC” are not actually macrophages.
  This came up because CHIKV RNA+ DC have the highest enrichment for interaction
  with CHIKV RNA+ macrophages.

<br>

<br>

---

```{r "setup", include = FALSE}
# Default chunk options
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo    = FALSE,
  dpi     = 250
)

# Set python environment
reticulate::use_condaenv("giotto", conda = "~/.local/bin/micromamba")

GiottoClass::set_giotto_python_path("giotto")

# Setup chunks
knitr::knit(here::here(params$template_dir, "setup.qmd"), output = "")

# Knit Xenium template
# * Objects storing strings should be evaluated before passing to template, and
#   should not be quoted
# * Objects storing other data structures, e.g., lists, must be evaluated within
#   the template and should be quoted
chunk_params <- list(
  file           = here(params$template_dir, "setup-xenium.qmd"),
  obj_prfx       = "xen",
  obj_dir        = params$xen_obj_dir,
  ref_obj        = params$xen_ref,
  data_dir       = params$xen_data_dir,
  data_regex     = params$xen_data_regex,
  sample_regex   = params$xen_sample_regex,
  sample_names   = "params$xen_sample_names",  # quote list objects
  sample_delim   = "-",
  custom_targets = params$xen_targets,
  key_targets    = "params$xen_key_targets",   # quote list objects
  top_types      = "Macrophages"
)

chunks <- do.call(knit_expand, chunk_params)

knit(text = chunks, output = "")
```

```{r "format data"}
# Cell to plot
plt_cells <- colnames(small)

# Sample levels
sam_lvls <- unname(params$xen_sample_names) %>%
  unlist() %>%
  unique()

# Key features
# * select based on median prediction confidence
# * make sure T cells are included
key_feats <- names(params$xen_key_targets)

# Select cell types to show in main figures
main_types <- xen@meta.data %>%
  group_by(cell_type) %>%
  summarize(pred_conf = median(pred_conf)) %>%
  arrange(desc(pred_conf)) %>%
  pull(cell_type)

main_types <- main_types[1:which(main_types == "DC")]

main_types <- unique(c(main_types, "T cells"))

xen <- xen %>%
  mutate_meta(
    mutate,
    cell_type_2 = if_else(
      cell_type %in% main_types,
      cell_type,
      "other"
    ),
    class_type = ifelse(
      cell_class == "high",
      cell_type,
      "CHIKV-"
    ),
    class_type_2 = if_else(
      cell_class == "low",
      "CHIKV-",
      cell_type_2
    )
  )

# Format plotting data
clmns <- c(
  ".cell_id", "sample", "fov", "cell_type",
  "nCount_Xenium", "nCount_Xenium_custom",
  "cell_class", "class_type",
  "cell_type_2", "class_type_2", "pred_conf",
  key_feats
)

xen_dat <- xen@meta.data %>%
  as_tibble(rownames = ".cell_id") %>%
  dplyr::select(
    all_of(clmns),
    any_of(c("treatment", "rep")),
    matches("^[xy]_cell$")
  ) %>%
  mutate(
    across(
      all_of(key_feats),
      ~ .x / (nCount_Xenium + nCount_Xenium_custom),
      .names = "frac_{.col}"
    )
  ) %>%
  rowwise() %>%
  mutate(
    nCount_key_feats = sum(!!!syms(key_feats))
  ) %>%
  ungroup() %>%
  mutate(
    frac_key_feats = nCount_key_feats / (nCount_Xenium + nCount_Xenium_custom)
  ) %>%
  ungroup() %>%
  mutate(
    sample     = fct_relevel(sample, sam_lvls),
    cell_class = fct_relevel(cell_class, c("low", "high")),
    fov_lab    = str_extract(fov, "[A-Z]$"),
    fov_lab    = str_c(sample, "\n", fov)
  )

# Set replicate if not already set
if (!"rep" %in% colnames(xen_dat)) {
  xen_dat <- xen_dat %>%
    group_by(sample) %>%
    mutate(rep = as.numeric(factor(fov))) %>%
    ungroup()
}

# Set FOV levels
fov_lvls <- xen_dat %>%
  arrange(sample) %>%
  distinct(fov, fov_lab)

fov_lvls <- set_names(fov_lvls$fov_lab, fov_lvls$fov)

fovs <- names(fov_lvls)

xen_dat <- xen_dat %>%
  mutate(fov = fct_relevel(fov, names(fov_lvls))) %>%
  arrange(fov) %>%
  mutate(fov_lab = fct_inorder(fov_lab))

# Format small data
small <- small %>%
  AddMetaData(FetchData(xen, c("cell_type_2", "class_type_2")))

small_dat <- xen_dat %>%
  filter(.cell_id %in% plt_cells)

# Select the top cell types with the most CHIKV+ cells
# * must have at least 20 CHIKV+ for the cell type to be included
chikv_types <- xen_dat %>%
  filter(treatment == "CHIKV", cell_class == "high") %>%
  mutate(cell_type = fct_reorder(cell_type, frac_key_feats, mean, .desc = TRUE)) %>%
  group_by(sample, cell_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  group_by(cell_type) %>%
  filter(min(n) >= 20) %>%
  summarize(n = sum(n)) %>%
  arrange(cell_type) %>%
  pull(n, cell_type)
```

```{r "theme"}
# ggplot2 themes
txt_pt1  <- 10
txt_pt2  <- 12
ttl_pt1  <- 14
ttl_pt2  <- 16
ln_pt    <- 0.5
ln_col   <- "grey85"
cir_size <- 3.5
sqr_size <- 4

txt_theme_1 <- theme(
  strip.text   = element_text(size = ttl_pt2),
  legend.text  = element_text(size = ttl_pt1),
  axis.title   = element_text(size = ttl_pt2),
  axis.text    = element_text(size = txt_pt2, color = "black"),
  legend.title = element_text(size = ttl_pt2),
)

line_theme <- theme(
  axis.line.x  = element_line(linewidth = ln_pt, color = ln_col),
  axis.line.y  = element_line(linewidth = ln_pt, color = ln_col),
  axis.ticks.x = element_line(linewidth = ln_pt, color = ln_col),
  axis.ticks.y = element_line(linewidth = ln_pt, color = ln_col)
)

base_theme <- djvdj_theme() +
  theme(
    plot.title       = element_text(face = "plain", size = ttl_pt2),
    strip.background = element_blank(),
    strip.text       = element_text(face = "plain")
  ) +
  txt_theme_1

umap_theme <- base_theme +
  theme(
    panel.border = element_rect(color = NA, fill = NA),
    axis.title   = element_blank(),
    axis.line.x  = element_blank(),
    axis.line.y  = element_blank(),
    axis.ticks.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text    = element_blank()
  )

umap_theme_2 <- umap_theme %+replace%
  theme(panel.border = element_rect(colour = ln_col, linewidth = ln_pt, fill = NA))

fade_0 <- "#FAFAFA"
fade_1 <- "#F0F0F0"
fade_2 <- "#D9D9D9"

# Colors
# Okabe Ito color palettes
ito_cols <- c(
  palette_OkabeIto[c(1:3, 5)], "#D7301F",
  palette_OkabeIto[c(4, 6)],   "#6A51A3",
  palette_OkabeIto[7],         "#875C04",
  "#065D43", "#F7AF34", "#00446E"
)

ito_cols <- ito_cols[3:length(ito_cols)] %>%
  darken(0.2) %>%
  c(ito_cols, ., "#686868", "#000000")

clrs <- ito_cols

# Cell type colors
typs <- unique(xen$cell_type)
typs <- typs[typs != "unassigned"]

typ_clrs <- set_names(
  clrs[seq_along(typs)],
  typs
)

typ_clrs <- typ_clrs %>%
  spruce_colors(difference = Inf, filter = "colorblind")

typ_clrs[c("CHIKV-", "low", "unassigned")] <- "grey85"
typ_clrs["other"]            <- "grey50"
typ_clrs["Macrophages"]      <- "#D7301F"

typ_clrs["Fibroblasts"] <- darken(typ_clrs["Fibroblasts"], 0.75)

# typ_clrs[c("LEC", "BEC", "cDC1")] <- c("#D7301F", "#0072B2", "#6A51A3")
# typ_clrs[c("LEC", "BEC", "cDC1")] <- c("#56B4E9", "#D7301F", "#6A51A3")
# typ_clrs[c("LEC", "cDC1", "Monocytes", "B cells")] <- c("#0072B2", "#6A51A3", "#D55E00", "#56B4E9")
# typ_clrs["Macrophages"] <- "black"

top_typ_clrs <- typ_clrs
top_typ_clrs[!names(top_typ_clrs) %in% top_types] <- "grey85"

top_typ_clrs_2 <- top_typ_clrs
top_typ_clrs_2[!names(top_typ_clrs_2) %in% top_types] <- "grey99"

# # Sample colors
# sam_clrs <- set_names(
#   c("#009E73", "#0072B2", "#56B4E9"),
#   c(sam_lvls, "Mock")
# )
```

```{r "QC", eval = FALSE}
# Format plotting data
feats <- c(
  "sample", "fov", "cell_type",
  "nCount_Xenium", "nCount_Xenium_custom"
)

DefaultAssay(xen) <- "Xenium_custom"

qc_dat <- xen %>%
  FetchData(c(feats, custom_targets)) %>%
  as_tibble(rownames = ".cell_id") %>%
  pivot_longer(all_of(custom_targets))

# Format heatmap data
heat_dat <- qc_dat %>%
  group_by(fov, name) %>%
  mutate(
    tot_counts = sum(nCount_Xenium + nCount_Xenium_custom)
  ) %>%
  group_by(name, fov, sample, tot_counts) %>%
  summarize(
    value   = sum(value),
    .groups = "drop"
  ) %>%
  mutate(
    frac   = value / tot_counts,
    fov    = fct_relevel(fov, names(fov_lvls)),
    sample = fct_relevel(sample, sam_lvls)
  )

# Tiles to outline
top_dat <- heat_dat %>%
  filter(frac >= 0.0001)

# Tiles to label
lab_dat <- heat_dat %>%
  filter(frac >= 0.0001) %>%
  mutate(
    lab = round(frac * 100, 2),
    lab = str_c(lab, "%")
  )

# Create heatmap
heat_dat %>%
  ggplot(aes(fov, name, fill = frac + 0.0001)) +
  geom_tile() +
  geom_tile(
    data = top_dat,
    color = "black"
  ) +
  geom_text(
    aes(color = NULL, fill = NULL, label = lab),
    data = lab_dat
  ) +
  facet_grid(~ sample, space = "free", scales = "free") +
  scale_fill_gradientn(colours = c("white", "#D7301F"), labels = label_percent(), trans = "log10") +
  scale_color_manual(values = c(`TRUE` = "black", `FALSE` = "white")) +
  base_theme +
  theme(
    # aspect.ratio       = n_distinct(heat_dat$fov) / n_distinct(heat_dat$name),
    strip.clip         = "off",
    strip.text.y.right = element_text(angle = 0, hjust = 0),
    axis.text.x        = element_text(angle = 90, vjust = 0.5, hjust = 1)
  )


# # Boxplots
# qc_dat %>%
#   ggplot(aes(cell_type, value + 1, fill = cell_type)) +
#   geom_boxplot(outlier.size = 0.1) +
#   facet_grid(name ~ fov, scales = "free") +
#   scale_y_log10() +
#   base_theme +
#   theme(
#     strip.clip         = "off",
#     strip.text.y.right = element_text(angle = 0),
#     axis.text.x        = element_text(angle = 90, vjust = 0.5, hjust = 1)
#   )
```

## Cell types

Cell types were annotated using 28 dpi CHIKV scRNA-seq data.
Annotated cell types are shown for tissue sections.

```{r "cell type sections", fig.width = 18, fig.height = 15}
# Legend labels
cell_n <- xen_dat %>%
  group_by(cell_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    pct   = round((n / sum(n)) * 100, 1),
    n_lab = case_when(
      n >= 10000 ~ str_c(round(n / 1000, 0), "k"),
      n >= 1000  ~ str_c(round(n / 1000, 1), "k"),
      TRUE       ~ label_comma()(n)
    ),
    n_lab = str_c(cell_type, " (", n_lab, ", ", pct, "%)")
  ) %>%
  arrange(desc(n))

lgnd_labs <- set_names(cell_n$n_lab, cell_n$cell_type)

# Create section plots
plts <- fovs %>%
  map(~ {
    p <- xen_dat %>%
      filter(fov == .x) %>%
      mutate(cell_type = fct_relevel(cell_type, cell_n$cell_type)) %>%
      arrange(cell_type) %>%
      ggplot(aes(x_cell, y_cell, color = cell_type)) +
      geom_point(size = 0.0001) +
      guides(color = guide_legend(override.aes = list(shape = 15, size = 4))) +
      scale_color_manual(values = typ_clrs) +
      labs(tag = params$xen_sample_names[[.x]]) +
      coord_fixed() +
      scale_y_reverse(expand = expansion(c(0.05, 0.15))) +
      umap_theme +
      theme(
        legend.position   = "none",
        plot.tag.position = c(0.15, 0.95),
        plot.tag          = element_text(size = ttl_pt2 * 1.5)
      )
    p
  })

plts <- plts %>%
  wrap_plots(
    nrow = 2,
    guides = "collect"
  )

# Create heatmaps
ht_dat <- xen_dat %>%
  group_by(sample) %>%
  mutate(n = n_distinct(.cell_id)) %>%
  group_by(sample, cell_type, n) %>%
  summarize(n_type = n_distinct(.cell_id), .groups = "drop") %>%
  mutate(
    pct_type = n_type / n,
    n_lab    = case_when(
      n_type >= 10000 ~ str_c(round(n_type / 1000, 0), "k"),
      n_type >= 1000  ~ str_c(round(n_type / 1000, 1), "k"),
      TRUE            ~ as.character(n_type)
    ),
    cell_type = fct_reorder(cell_type, pct_type, .desc = TRUE)
  )

ht <- ht_dat %>%
  ggplot(aes(sample, pct_type, fill = cell_type)) + 
  geom_col() +
  geom_text(
    aes(label = n_lab, fill = NULL),
    vjust = -0.1,
    size = txt_pt1 * 1.5 / .pt
  ) +
  facet_wrap(~ cell_type, scale = "free_y") +
  scale_fill_manual(values = typ_clrs) +
  scale_y_continuous(labels = label_percent(), breaks = breaks_pretty(n = 3), expand = expansion(c(0.05, 0.25))) +
  labs(y = "% of total cells") +
  base_theme +
  theme(
    legend.position = "none",
    axis.title.x    = element_blank(),
    axis.title.y    = element_text(size = txt_pt2 * 1.5),
    axis.text.x     = element_text(size = txt_pt1 * 1.5, angle = 45, hjust = 1),
    axis.text.y     = element_text(size = txt_pt1 * 1.2),
    
    strip.text      = element_text(size = ttl_pt2 * 1.5)
  )

# Create final figure
fig <- plot_grid(
  plts, ht,
  ncol = 1,
  rel_heights = c(1, 0.7)
)

fig
```

<br>

### {.tabset .tabset-pills}

Tissue sections are shown with only the top cell types labeled
(assessed based on annotation confidence shown below).

```{r "top cell types sections", fig.width = 16, fig.height = 8, results = "asis", out.height = "70%"}
fovs %>%
  walk(~ {
    cat("\n\n####", params$xen_sample_names[[.x]], "\n\n")
    
    fig <- xen_dat %>%
      .plot_full_fov(
        fov   = .x,
        color = "cell_type_2",
        trace = "Macrophages",
        clrs  = typ_clrs,
        size       = 0.5,
        trace_size = 0.5,
        trace_clr  = NA
      )
    
    print(fig)
    cat("\n\n<br>\n\n<br>\n\n")
  })
```

<br>

### {.tabset .tabset-pills}

High resolution images show all cells (left) or all cells with
`r str_flatten(top_types, collapse = ", ", last = " and ")`
highlighted (right).

```{r "hi res cell types sections", fig.width = 45, fig.height = 10, results = "asis", out.height = "70%"}
fovs %>%
  walk(~ {
    cat("\n\n####", params$xen_sample_names[[.x]], "\n\n")
    
    fig <- xen_dat %>%
      .plot_full_fov(
        fov    = .x,
        color  = "cell_type_2",
        trace  = list("NONE", "Macrophages"),
        clrs   = typ_clrs,
        show_n = FALSE
      ) %>%
      plot_grid(
        plotlist = .,
        nrow  = 1,
        align = "h",
        axis  = "tb"    
      )
    
    print(fig)
    cat("\n\n<br>\n\n<br>\n\n")
  })
```

<br>

### {.tabset .tabset-pills}

Heatmaps show the spatial distribution of cell types for each sample.

```{r "xenium cell type sections", fig.width = 14, fig.height = 12, results = "asis"}
# Set cell types to plot
xen_typs <- xen_dat$cell_type %>%
  table() %>%
  sort(decreasing = TRUE) %>%
  names()

xen_typs <- xen_typs[xen_typs != "unassigned"]

# Create heatmaps
fovs %>%
  walk(~ {
    cat("\n\n####", params$xen_sample_names[[.x]], "\n\n")
    
    fig <- xen_dat %>%
      filter(fov == .x) %>%
      .create_cell_heatmap(groups = xen_typs, n_bins = 100)
    
    print(fig)
    cat("\n\n<br>\n\n<br>\n\n")
  })
```

### Annotation confidence

The per-cell prediction confidence for cell type annotations is shown below.
Scores approaching 1 indicate higher confidence.
Some unassigned cells have a high prediction confidence since the reference
includes cells that were unassigned in the scRNA-seq data.

```{r "pred confidence", fig.width = 9, fig.height = 4, out.width = "70%"}
dat <- xen_dat %>%
  mutate(
    cell_type = fct_reorder(cell_type, pred_conf, .desc = TRUE)
  )

dat %>%
  ggplot(aes(cell_type, pred_conf, fill = cell_type)) +
  geom_hline(yintercept = 0.5, linetype = 2, linewidth = 0.5) +
  geom_boxplot(outlier.size = 0.2) +
  facet_wrap(~ sample) +
  scale_fill_manual(values = typ_clrs) +
  labs(y = "prediction confidence") +
  base_theme +
  theme(
    legend.position = "none",
    axis.text.x     = element_text(angle = 45, hjust = 1),
    axis.title.x    = element_blank()
  )
```

<br>

### {.tabset .tabset-pills}

Expression of select marker genes is shown below.
The % of cells expressing each gene is shown above each boxplot.

```{r "cell type markers"}
gns <- c(
  "Ptprc",
  "Cd3d", "Cd19",            # B/T cells
  "Mrc1", "Folr2", "Cx3cr1", # macrophages
  "Aqp1",
  "Serpinf1",                # fibroblasts
  "Cd34", "Pecam1",          # endothelial cells
  "Itgax",                   # DCs
  "Csf3r",                   # neutrophil
  # "Irf7",                  # monocytes
  "Acta2", "Itga7",          # PvCs
  "Ttn"                      # skeletal muscle
)

plts <- sam_lvls %>%
  set_names() %>%
  map(~ {
    sam <- .x
    
    plt <- gns %>%
      map(~ {
        dat <- xen %>%
          FetchData(c("sample", "cell_type", .x)) %>%
          filter(sample == sam)
        
        n_expr <- dat %>%
          group_by(cell_type) %>%
          summarize(
            med     = median(!!sym(.x)),
            n       = n(),
            n_expr  = sum(!!sym(.x) > 0),
            .groups = "drop"
          ) %>%
          mutate(
            pct_expr = n_expr / n,
            pct_lab  = label_percent()(round(pct_expr, 2))
          )
        
        lvls <- n_expr %>%
          arrange(desc(med), desc(pct_expr)) %>%
          pull(cell_type)
        
        dat %>%
          plot_violin(
            .x,
            cluster_col   = "cell_type",
            plot_colors   = typ_clrs,
            plot_lvls     = lvls,
            method        = "boxplot",
            color         = "black",
            alpha         = 1,
            outlier.size  = 0.5,
            outlier.alpha = 1
          ) +
          scale_y_continuous(expand = expansion(c(0.05, 0.2))) +
          geom_text(
            aes(cell_type, Inf, label = pct_lab, fill = NULL),
            data = n_expr,
            vjust = 1.5,
            color = "black",
            size  = 8 / .pt
          )
      }) %>%
      plot_grid(
        plotlist = .,
        ncol     = 3,
        align    = "vh",
        axis     = "trbl"
      )
  })
```

```{r "cell type markers 2", fig.width = 15, fig.height = 15, results = "asis"}
plts %>%
  iwalk(~ {
    cat("\n\n####", .y, "\n\n")
    print(.x)
    cat("\n\n<br>\n\n<br>\n\n")
  })
```

### Cell type markers {.tabset .tabset-pills}

The top 40 marker genes identified for each cell type are shown below for
cells from mock-infected samples.
Differentially expressed genes were filtered to only include upregulated genes
with a minimum adjusted p value <0.05.
Arrows indicate the cell type used to identify each set of markers.

```{r "identify degs"}
# Prepare object
xen <- xen %>%
  PrepSCTFindMarkers(assay = "SCT")

Idents(xen) <- xen$cell_type

# Identify DEGs
degs <- xen %>%
  find_conserved_markers(
    assay        = "SCT",
    rep_var      = "sample",
    fc_min       = NULL,
    fc_range     = c(0, Inf),
    p_adj_method = "BH",
    file_path    = here(params$table_dir, "cell_type_markers.tsv.gz") 
  )

# Filter DEGs
pct_clmns <- syms(str_c(sam_lvls, "_pct.1"))

top_degs <- degs %>%
  filter(
    all_same_direction,
    direction == "up",
    min_p_adj < 0.05
  ) %>%
  rowwise() %>%
  mutate(min_pct_1 = min(!!!pct_clmns)) %>%
  ungroup()
```

```{r "cell type marker plots", fig.width = 15, fig.height = 8, results = "asis"}
lvls <- top_degs %>%
  group_by(ident_1) %>%
  summarize(n = n(), .groups = "drop") %>%
  arrange(desc(n)) %>%
  pull(ident_1)

marks <- top_degs %>%
  arrange(desc(min_pct_1), max_p_adj, min_p_adj) %>%
  group_by(ident_1) %>%
  slice(1:40) %>%
  split(.$ident_1) %>%
  map(pull, gene)

marks <- marks[lvls]

dat <- xen %>%
  subset(treatment == "mock")

marks %>%
  iwalk(~ {
    cat("\n\n####", .y, "\n\n")
    
    plt <- dat %>%
      create_feat_boxes(
        grp_column    = "cell_type",
        feats         = .x,
        box_colors    = typ_clrs,
        highlight_grp = .y,
        y_ttl         = "normalized expression"
      )
    
    print(plt)
    cat("\n\n---\n\n<br>\n\n<br>\n\n")
  })
```

## CHIKV infection

### Macrophages harbor CHIKV RNA {.tabset .tabset-pills}

CHIKV+ cells (>0 total CHIKV counts) are shown for tissue sections.

```{r "CHIKV+ cell types", fig.width = 17, fig.height = 11}
.create_class_figure(
  bar_dat       = xen_dat,
  plot_dat      = xen,
  class_clmn    = "cell_class",
  class         = "high",
  celltype_clmn = "cell_type_2",
  clrs          = typ_clrs,
  top_types     = top_types,
  nrow          = 3,
  y_ttl         = "# CHIKV+ cells",
  alph          = rep(1, n_distinct(xen_dat$rep))
)
```

<br>

### {.tabset .tabset-pills}

Tissue sections are shown with only the top cell types
(based on annotation confidence) labeled for CHIKV+ cells.

```{r "top cell types CHIKV+ sections", fig.width = 16, fig.height = 8, results = "asis", out.height = "70%"}
fovs %>%
  walk(~ {
    cat("\n\n####", params$xen_sample_names[[.x]], "\n\n")
    
    fig <- xen_dat %>%
      .plot_full_fov(
        fov   = .x,
        color = "class_type_2",
        trace = "Macrophages",
        clrs  = typ_clrs,
        size       = 1,
        trace_size = 1,
        trace_clr  = NA
      )
    
    print(fig)
    cat("\n\n<br>\n\n<br>\n\n")
  })
```

### {.tabset .tabset-pills}

High-resolution images show CHIKV counts (left) and CHIKV+ cells (right)
for each tissue section.

```{r, fig.width = 45, fig.height = 10, results = "asis", out.height = "70%"}
fovs %>%
  walk(~ {
    cat("\n\n####", params$xen_sample_names[[.x]], "\n\n")
    
    plt <- xen_dat %>%
      .plot_full_fov(
        fov   = .x,
        color = "class_type_2",
        clrs  = typ_clrs,
        trace = "NONE"
      )
    
    sig <- xen_dat %>%
      .plot_full_fov_signal(
        fov    = .x,
        color  = "nCount_key_feats",
        clrs   = c("white", "#D7301F"),
        ttl    = "CHIKV\ncounts + 1",
        pseudo = 1,
        trans  = "log10"
      )
    
    fig <- plot_grid(
      sig, plt,
      nrow  = 1,
      align = "h",
      axis  = "tb" 
    )
      
    print(fig)
    cat("\n\n")
  })
```

### {.tabset .tabset-pills}

Heatmaps show the spatial distribution of CHIKV+ cells for each cell type.
Only cell types with at least 20 CHIKV+ cells for all replicates are shown.

```{r "CHIKV+ heatmap sections", fig.width = 14, fig.height = 10, results = "asis"}
fovs %>%
  walk(~ {
    cat("\n\n####", params$xen_sample_names[[.x]], "\n\n")
    
    dat <- xen_dat %>%
      filter(fov == .x)
    
    fig <- dat %>%
      .create_cell_heatmap(
        count_dat = filter(., cell_class == "high"),
        groups    = names(sort(chikv_types, decreasing = TRUE)),
        n_bins    = 75,
        n_col     = 4
      )
    
    print(fig)
    cat("\n\n<br>\n\n<br>\n\n")
  })
```

###

The fraction of total counts originating from CHIKV probes is shown below
for all cells (top) or CHIKV+ cells (bottom).
Cell types with <20 cells are not shown.
The number of cells identified for each cell type is shown above each boxplot.

```{r "CHIKV COUNTS CELL TYPES", fig.width = 15, fig.height = 6}
# Format CHIKV cell type data
type_dat <- xen_dat %>%
  group_by(sample, cell_type) %>%
  filter(n() > 20) %>%
  ungroup() %>%
  mutate(
    cell_type = fct_reorder(cell_type, frac_key_feats, mean, .desc = TRUE),
    sample = fct_relevel(sample, sam_lvls)
  )

# Format N labels
n_dat <- type_dat %>%
  group_by(sample, cell_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n_lab = if_else(
      n > 1000,
      str_c(round(n / 1000, 1), "k"),
      label_comma()(n)
    )
  )

# Fraction CHIKV reads boxplots
pseudo <- 0.01

type_dat %>%
  ggplot(aes(cell_type, frac_key_feats + pseudo, fill = cell_type, alpha = as.character(rep))) +
  geom_boxplot(
    outlier.size = 0.1,
    outlier.alpha = 1
  ) +
  geom_text(
    aes(y = Inf, label = n_lab, alpha = NULL, fill = NULL),
    data = n_dat,
    size = txt_pt1 / .pt,
    vjust = 1.25
  ) +
  facet_wrap(~ sample) +
  scale_y_log10(expand = expansion(c(0.05, 0.2)), labels = label_percent()) +
  scale_fill_manual(values = typ_clrs) +
  scale_alpha_manual(values = c("1" = 1, "2" = 0.75, "3" = 0.5, "4" = 0.25)) +
  guides(fill = "none", alpha = "none") +
  labs(title = "all cells", y = str_c("% CHIKV counts + ", pseudo * 100)) +
  base_theme +
  theme(
    panel.background = element_rect(fill = NA, color = "grey75", linewidth = ln_pt),
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )
```

```{r "CHIKV SIGNAL CELL TYPES", fig.width = 15, fig.height = 4}
# Format N labels
dat <- xen_dat %>%
  filter(treatment == "CHIKV", cell_class == "high") %>%
  group_by(sample, cell_type) %>%
  filter(n() >= 20) %>%
  ungroup() %>%
  mutate(
    cell_type = fct_reorder(cell_type, frac_key_feats, mean, .desc = TRUE),
    sample = fct_relevel(sample, sam_lvls)
  )

n_dat <- dat %>%
  group_by(sample, cell_type) %>%
  summarize(n = n(), .groups = "drop") %>%
  mutate(
    n_lab = if_else(
      n > 1000,
      str_c(round(n / 1000, 1), "k"),
      label_comma()(n)
    )
  )

# Fraction CHIKV reads boxplots
dat %>%
  ggplot(aes(cell_type, frac_key_feats, fill = cell_type)) +
  geom_boxplot(
    outlier.size = 0.1,
    outlier.alpha = 1
  ) +
  geom_text(
    aes(y = Inf, label = n_lab, alpha = NULL, fill = NULL),
    data = n_dat,
    size = txt_pt1 / .pt,
    vjust = 1.25
  ) +
  facet_wrap(~ sample) +
  scale_y_log10(expand = expansion(c(0.05, 0.2)), labels = label_percent()) +
  scale_fill_manual(values = typ_clrs) +
  guides(fill = "none", alpha = "none") +
  labs(title = "CHIKV+ cells", y = str_c("% CHIKV counts")) +
  base_theme +
  theme(
    aspect.ratio = 0.7,
    panel.background = element_rect(fill = NA, color = "grey75", linewidth = ln_pt),
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )
```

<br>

<br>

### CHIKV signals are higher for CHIKV 3' probes

The % CHIKV+ cells is shown for each cell type.
The number of CHIKV+ cells is shown above each bar.

```{r "CHIKV+ bargraphs PAPER", fig.width = 8, fig.height = 4, out.width = "70%"}
frac_clmns <- str_c("frac_", key_feats)

# Format bargraph data
bar_dat <- xen_dat %>%
  filter(
    treatment == "CHIKV",
    cell_type %in% c("Macrophages", "Fibroblasts")
  ) %>%
  group_by(sample, cell_type) %>%
  mutate(n = n()) %>%
  pivot_longer(all_of(key_feats)) %>%
  ungroup() %>%
  group_by(sample, cell_type, name, n) %>%
  summarize(
    n_pos   = sum(value > 0),
    .groups = "drop"
  ) %>%
  mutate(
    frac_pos = n_pos / n,
    across(
      c(n, n_pos),
      ~ if_else(.x >= 1000, str_c(round(.x / 1000, 1), "k"), as.character(.x))
    ),
    n_lab     = str_c(n_pos, "/", n),
    name      = fct_relevel(name, key_feats),
    cell_type = fct_reorder(cell_type, frac_pos, .desc = TRUE)
  )

# Plot labels
lbs <- c(
  "CHIKV_sgRNA" = "CHIKV 3'",
  "CHIKV_5"     = "CHIKV 5'",
  "CHIKV_neg"   = "CHIKV neg"
)

# Create bargraphs
bar_dat %>%
  filter(cell_type %in% c("Macrophages", "Fibroblasts")) %>%
  ggplot(aes(cell_type, frac_pos, fill = cell_type, group = sample)) +
  geom_col(position = position_dodge(width = 0.95)) +
  geom_text(
    aes(label = n_lab, fill = NULL),
    angle = 90, hjust = 0,
    position = position_dodge(width = 0.95)
  ) +
  facet_wrap(~ name, labeller = as_labeller(lbs)) +
  scale_fill_manual(values = typ_clrs) +
  scale_y_continuous(expand = expansion(c(0.05, 0.35)), labels = label_percent()) +
  labs(y = "% CHIKV+ cells") +
  base_theme +
  theme(
    legend.position = "none",
    axis.title.x    = element_blank()
  )
```

```{r "CHIKV+ bargraphs", fig.width = 10, fig.height = 4, out.width = "70%"}
frac_clmns <- str_c("frac_", key_feats)

bar_dat <- xen_dat %>%
  filter(treatment == "CHIKV") %>%
  group_by(cell_type) %>%
  mutate(n = n()) %>%
  pivot_longer(all_of(key_feats)) %>%
  ungroup() %>%
  group_by(cell_type, name, n) %>%
  summarize(
    n_pos   = sum(value > 0),
    .groups = "drop"
  ) %>%
  mutate(
    frac_pos = n_pos / n,
    across(
      c(n, n_pos),
      ~ if_else(.x >= 1000, str_c(round(.x / 1000, 1), "k"), as.character(.x))
    ),
    n_lab     = str_c(n_pos, "/", n),
    name      = fct_relevel(name, key_feats),
    cell_type = fct_reorder(cell_type, frac_pos, .desc = TRUE)
  )

# Create bargraphs
bar_dat %>%
  ggplot(aes(cell_type, frac_pos, fill = cell_type)) +
  geom_col() +
  geom_text(
    aes(label = n_lab, fill = NULL),
    angle = 90, hjust = 0
  ) +
  facet_wrap(~ name) +
  scale_fill_manual(values = typ_clrs) +
  scale_y_continuous(expand = expansion(c(0.05, 0.25)), labels = label_percent()) +
  labs(y = "% CHIKV+ cells") +
  base_theme +
  theme(
    legend.position = "none",
    axis.text.x     = element_text(angle = 45, hjust = 1),
    axis.title.x    = element_blank()
  )
```

```{r "CHIKV+ bargraphs DATA CLUB", fig.width = 10, fig.height = 4, out.width = "70%", eval = FALSE}
frac_clmns <- str_c("frac_", key_feats)

dat <- xen_dat %>%
  filter(treatment == "CHIKV") %>%
  group_by(cell_type) %>%
  summarize(
    n = n(),
    n_pos = sum(cell_class == "high"),
    .groups = "drop"
  ) %>%
  
  # filter(n_pos > 50) %>%
  
  mutate(
    frac_pos = n_pos / n,
    across(
      c(n, n_pos),
      ~ if_else(.x >= 1000, str_c(round(.x / 1000, 1), "k"), as.character(.x))
    ),
    n_lab     = str_c(n_pos, "/", n),
    cell_type = fct_reorder(cell_type, frac_pos, .desc = TRUE)
  )
  
# Create bargraphs
brs <- dat %>%
  ggplot(aes(cell_type, frac_pos, fill = cell_type)) +
  geom_col() +
  geom_text(
    aes(label = n_lab, fill = NULL),
    angle = 90, hjust = 0
  ) +
  scale_fill_manual(values = typ_clrs) +
  scale_y_continuous(expand = expansion(c(0.05, 0.25)), labels = label_percent()) +
  labs(y = "% CHIKV+ cells") +
  base_theme +
  theme(
    legend.position = "none",
    axis.text.x     = element_text(angle = 45, hjust = 1),
    axis.title.x    = element_blank()
  )

# Create boxplots
bxs <- xen_dat %>%
  filter(
    treatment == "CHIKV",
    cell_class == "high"
  ) %>%
  group_by(cell_type) %>%
  filter(n_distinct(.cell_id) > 50) %>%
  ungroup() %>%
  mutate(
    # cell_type = fct_relevel(cell_type, levels(dat$cell_type)),
    cell_type = fct_reorder(cell_type, frac_key_feats, median, .desc = TRUE)
  ) %>%
  ggplot(aes(cell_type, frac_key_feats, fill = cell_type)) +
  geom_boxplot(outlier.size = 0.1, outlier.alpha = 1) +
  
  # facet_wrap(~ rep) +
  
  scale_fill_manual(values = typ_clrs) +
  scale_y_continuous(trans = "log10", labels = label_percent()) +
  labs(y = "% CHIKV counts") +
  base_theme +
  theme(
    legend.position = "none",
    axis.text.x     = element_text(angle = 45, hjust = 1),
    axis.title.x    = element_blank()
  )

# Save final figure
plt <- plot_grid(
  brs, NULL, bxs,
  nrow  = 1,
  align = "h",
  axis  = "tb",
  rel_widths = c(1, 0.2, 1)
)

ggsave(
  "chikv_quant.png",
  path   = file.path("~/Dropbox/Ryan/Documents/Hesselberth_lab/Meetings/2025-01-28_data_club"),
  plot   = plt,
  device = "png",
  bg     = "white", 
  width  = 10,
  height = 4,
  dpi    = 600,
)
```

<br>

Total counts assigned to 3' (sgRNA), 5' (nsP genes), and
negative strand CHIKV probes is shown for CHIKV+ cells.
The number of cells plotted is shown above each boxplot.

```{r "CHIKV SIGNAL TYPES", fig.width = 15, fig.height = 4}
# CHIKV signal colors
key_nms <- set_names(
  key_feats,
  str_c("frac_", key_feats)
)

chikv_clrs <- set_names(c("#D7301F", "#0072B2", "#009E73"), names(key_nms))

# Create boxplots
plts <- names(chikv_types[1:4]) %>%
  map(~ {
    dat <- xen_dat %>%
      filter(cell_type == .x, treatment == "CHIKV", cell_class == "high") %>%
      pivot_longer(all_of(key_feats))
    
    clrs <- set_names(
      unname(chikv_clrs),
      key_feats
    )
    
    dat <- dat %>%
      mutate(name = fct_relevel(name, key_feats))
    
    y_ttl <- "CHIKV counts + 1"
    
    n_dat <- dat %>%
      group_by(sample) %>%
      summarize(n = n_distinct(.cell_id), .groups = "drop") %>%
      mutate(n_lab = label_comma()(n))
    
    dat %>%
      ggplot(aes(sample, value + 1, fill = name)) +
      geom_boxplot(
        outlier.size = 0.1,
        outlier.alpha = 1
      ) +
      geom_text(
        aes(y = Inf, label = n_lab, fill = NULL),
        data  = n_dat,
        color = "black",
        vjust = 1.25
      ) +
      scale_y_log10(expand = expansion(c(0.05, 0.2))) +
      scale_fill_manual(values = clrs) +
      labs(title = str_c("CHIKV+ ", .x), y = y_ttl) +
      base_theme +
      theme(
        aspect.ratio     = 1,
        plot.margin      = margin(15, 15, 15, 30, "pt"),
        panel.background = element_rect(fill = NA, color = "grey75", linewidth = ln_pt),
        legend.title     = element_blank(),
        axis.title.x     = element_blank(),
        axis.text.x      = element_text(angle = 45, hjust = 1)
      )
  })

wrap_plots(
  plts,
  guides = "collect",
  nrow = 1
)
```

<br>

The fraction of CHIKV+ cells with detectable 3' (sgRNA), 5' (nsP genes), and
negative strand signal is shown for each sample.
The CHIKV+ class indicates which probes were detectable within the cell, e.g.,
"sgRNA/5'/neg" had signal for all three regions and "sgRNA" only had
signal for sgRNA probes.
Only cell types with >100 total CHIKV+ cells are shown.

```{r "CHIKV SIGNAL GROUPS PAPER", fig.width = 6, fig.height = 4, out.width = "70%"}
# Set CHIKV groups
xen_dat <- xen_dat %>%
  mutate(
    chikv_class = case_when(
      CHIKV_sgRNA > 0 & CHIKV_5 > 0 & CHIKV_neg > 0 ~ "3'/5'/neg",
      CHIKV_sgRNA > 0 & CHIKV_5 > 0                 ~ "3'/5'",
      CHIKV_sgRNA > 0 & CHIKV_neg > 0               ~ "3'/neg",
      CHIKV_sgRNA > 0                               ~ "3'",
      CHIKV_5 > 0 & CHIKV_neg > 0                   ~ "5'/neg",
      CHIKV_5 > 0                                   ~ "5'",
      CHIKV_neg > 0                                 ~ "neg",
      cell_class == "low"                           ~ "CHIKV-",
      TRUE                                          ~ "other"
    )
  )

# Format plot data
grp_dat <- xen_dat %>%
  filter(
    treatment  == "CHIKV",
    cell_class == "high"
  ) %>%
  mutate(chikv_class = fct_infreq(chikv_class)) %>%
  group_by(cell_type) %>%
  filter(n_distinct(.cell_id) > 100) %>%
  ungroup() %>%
  mutate(cell_type = fct_infreq(cell_type))

n_dat <- grp_dat %>%
  group_by(sample, cell_type) %>%
  summarize(n = n_distinct(.cell_id), .groups = "drop") %>%
  mutate(
    n_lab = if_else(n >= 1000, str_c(round(n / 1000, 1), "k"), as.character(n))
  )

# Create bargraphs
lvls <- c("Macrophages", "Fibroblasts")

grp_dat %>%
  filter(cell_type %in% lvls) %>%
  mutate(cell_type = fct_relevel(cell_type, lvls)) %>%
  
  ggplot(aes(sample, fill = chikv_class)) +
  geom_bar(position = "fill") +
  geom_text(
    aes(y = 1, label = n_lab, fill = NULL),
    data = ~ filter(n_dat, cell_type %in% c("Macrophages", "Fibroblasts")),
    vjust = 0
  ) +
  facet_wrap(~ cell_type, ncol = 4) +
  scale_fill_manual(values = clrs) +
  scale_y_continuous(expand = expansion(c(0.05, 0.15)), breaks = c(0, 0.5, 1)) +
  labs(y = "fraction of CHIKV+ cells") +
  guides(fill = guide_legend(title = "CHIKV+ class")) +
  base_theme +
  theme(
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )
```

```{r "CHIKV SIGNAL GROUPS", fig.width = 10, fig.height = 5, out.width = "70%"}
# Create bargraphs
grp_dat %>%
  ggplot(aes(sample, fill = chikv_class)) +
  geom_bar(position = "fill") +
  geom_text(
    aes(y = 1, label = n_lab, fill = NULL),
    data = n_dat,
    vjust = 0
  ) +
  facet_wrap(~ cell_type, ncol = 4) +
  scale_fill_manual(values = clrs) +
  scale_y_continuous(expand = expansion(c(0.05, 0.15)), breaks = c(0, 0.5, 1)) +
  labs(y = "fraction of CHIKV+ cells") +
  guides(fill = guide_legend(title = "CHIKV+ class")) +
  base_theme +
  theme(
    axis.title.x = element_blank(),
    axis.text.x  = element_text(angle = 45, hjust = 1)
  )
```

---

<br>

<br>

## Cell proximity enrichment {.tabset .tabset-pills}

```{r "giotto setup"}
create_giotto <- !file.exists(here("results/prox.tsv.gz"))

.create_prox_bars <- function(df_in, cell_type = "Macrophages", clrs = c(
    `CHIKV+` = typ_clrs[["Macrophages"]],
    `CHIKV-` = "#53B3E1",
    none     = typ_clrs[["low"]]
  )) {
  
  re  <- str_replace_all(cell_type, "\\(", "\\\\(")
  re  <- str_replace_all(re, "\\)", "\\\\)")
  re  <- str_c(re, " \\(CHIKV\\+\\)")
  
  dat <- df_in %>%
    arrange(sample, enrichm) %>%
    mutate(
      highlight = grepl(re, unified_int),
      
      pos = case_when(
        highlight & significant & grepl("CHIKV\\+", unified_int) & grepl("CHIKV\\-", unified_int) ~ "CHIKV-",
        highlight & significant ~ "CHIKV+",
        TRUE ~ "none"
      ),
      interaction = str_remove(unified_int, str_c("[-]*", re, "[-]*")),
      interaction = str_remove(interaction, "[ ]*\\(CHIKV[-+]\\)[ ]*"),
      x = str_c(sample, "-", unified_int),
      x = fct_inorder(x)
    )
  
  # Create bargraphs
  res <- dat %>%
    ggplot(aes(x, enrichm, fill = pos)) +
    geom_col(width = 1) +
    geom_text_repel(
      aes(label = interaction),
      data    = ~ filter(.x, highlight & significant & enrichm > 0),
      nudge_y = 1,
      hjust   = 1,
      size    = 7/ .pt,
      min.segment.length = 0
    ) +
    geom_text_repel(
      aes(label = interaction),
      data    = ~ filter(.x, highlight & significant & enrichm < 0),
      nudge_y = -1,
      hjust   = 1,
      size    = 7 / .pt,
      min.segment.length = 0
    ) +
    facet_wrap(~ sample, nrow = 1, scales = "free") +
    scale_fill_manual(values = clrs, str_c("interactions with CHIKV+ ", cell_type)) +
    labs(
      y = "cell-cell interaction\nenrichment",
      x = "cell-cell interactions"
    ) +
    base_theme +
    theme(
      legend.position = "bottom",
      legend.title.position = "top",
      axis.title.x    = element_text(vjust = 2),
      axis.text.x     = element_blank(),
      axis.ticks.x    = element_blank()
    )
  
  res
}
```

```{r "giotto object", include = FALSE, eval = create_giotto}
# Remove unneeded objects
rm(xen)
rm(small)
gc()

# Load CHIKV samples
gio_dirs <- dir(
  here(data_dir),
  pattern = data_regex,
  full.names = TRUE
)

names(gio_dirs) <- gio_dirs %>%
  str_extract(sample_regex)

nms <- params$xen_sample_names
nms <- names(nms[grepl("^CHIKV", nms)])

gio_dirs <- gio_dirs[nms]

# Perform cell proximity analysis
# * not enough memory to run all together, have to run one by run and reload results
prox_res <- gio_dirs %>%
  imap(~ {
    # Create Giotto object
    # * store custom targets as separate assay
    nm <- str_c("_", params$xen_sample_names[[.y]])
    
    meta <- xen_dat %>%
      filter(fov == .y) %>%
      select(cell_ID = .cell_id, cell_type, cell_class) %>%
      mutate(
        cell_class = recode(cell_class, high = "CHIKV+", low = "CHIKV-"),
        cell_type  = str_c(cell_type, " (", cell_class, ")"),
        cell_ID    = str_remove(cell_ID, "^[0-9A-Z\\-]+_")
      ) %>%
      select(-cell_class)
    
    prox <- meta %>%
      run_giotto(
        xen_dir     = .x,
        clmn        = "cell_type",
        file_dir    = "results",
        file_prefix = nm
      )
    
    gc()
    
    prox
  })

# Format proximity enrichment results
prox_df <- params$xen_sample_names[names(gio_dirs)] %>%
  unname() %>%
  map_dfr(~ {
    obj <- qload(here(str_c("results/prox_", .x, ".qs")))
    
    obj$enrichm_res %>%
      as_tibble() %>%
      mutate(sample = .x)
  })

prox_df %>%
  write_tsv(here("results/prox.tsv.gz"))
```

```{r "T cell proximity", include = FALSE, eval = FALSE}
prox_res <- gio_dirs %>%
  imap(~ {
    # Create Giotto object
    # * store custom targets as separate assay
    nm <- str_c("_", params$xen_sample_names[[.y]], "_bulk")
    
    meta <- xen_dat %>%
      filter(fov == .y) %>%
      select(cell_ID = .cell_id, cell_type) %>%
      mutate(cell_ID = str_remove(cell_ID, "^[0-9A-Z\\-]+_"))
    
    prox <- meta %>%
      run_giotto(
        xen_dir     = .x,
        clmn        = "cell_type",
        file_dir    = "results",
        file_prefix = nm
      )
    
    gc()
    
    prox
  })

# Format proximity enrichment results
prox_df <- params$xen_sample_names[names(gio_dirs)] %>%
  unname() %>%
  map_dfr(~ {
    obj <- qload(here(str_c("results/prox_", .x, "_bulk.qs")))
    
    obj$enrichm_res %>%
      as_tibble() %>%
      mutate(sample = .x)
  })

prox_df %>%
  write_tsv(here("results/prox_bulk.tsv.gz"))
```

Cell-cell interaction enrichment is shown for each biological replicate for
CHIKV+ macrophages.
Each highlighted bar represents a cell type that showed increased
(positive enrichment) or decreased interaction with CHIKV+ macrophages.
Bars are colored based on whether the interacting cells were positive or
negative for CHIKV RNA (CHIKV+, CHIKV-).
Interactions are only shown for cell types with at least 20 CHIKV+ cells
identified in all three biological replicates.
For example, a blue bar labeled as "Macrophages" would represent the interaction
between CHIKV+ and CHIKV- macrophages.
Grey bars represent cell-cell interactions that did not include CHIKV+
macrophages.

```{r "mac giotto proximity", fig.width = 15, fig.height = 3.75}
# Load data
prox_df <- read_tsv(here("results/prox.tsv.gz")) %>%
  mutate(
    significant = (enrichm > 0 & p.adj_higher < 0.05) | (enrichm < 0 & p.adj_lower < 0.05)
  ) %>%
  filter(
    type_int == "hetero",
    significant,
    !is.na(type_int)
  )

# Filter cell types
prox_df <- prox_df %>%
  separate(
    unified_int,
    into = c("cell_type_1", "cell_type_2"),
    sep = "--",
    remove = FALSE
  ) %>%
  mutate(across(
    starts_with("cell_type_"),
    ~ str_remove(.x, " \\(CHIKV[-+]\\)$")
  )) %>%
  filter(
    cell_type_1 %in% names(chikv_types) &
      cell_type_2 %in% names(chikv_types)
  )

# Create proximity bargraph
prox_df %>%
  .create_prox_bars("Macrophages")
```

<br>

Cell-cell interaction enrichment is shown for other CHIKV+ cell types as described above.

```{r "other giotto proximity", fig.width = 15, fig.height = 3.75, results = "asis"}
typs <- xen_dat %>%
  filter(cell_type != "Macrophages", cell_class == "high") %>%
  pull(cell_type) %>%
  table() %>%
  sort(decreasing = TRUE)

typs <- typs[names(typs) %in% names(chikv_types)]

typs %>%
  iwalk(~ {
    cat("\n\n###", .y, "\n\n")
    cat(label_comma()(.x), "CHIKV+ cells were identified for", .y)
    
    plt <- prox_df %>%
      .create_prox_bars(.y)
    
    print(plt)
    cat("\n\n---\n\n<br>\n\n<br>\n\n")
  })
```

## Session info

```{r "session info"}
sessionInfo()
```

```{r "xenium metadata for geo", eval = FALSE}
clmns <- c(
  "cell_id",
  "fov",
  "treatment",
  "rep",
  "nCount_Xenium",
  "nFeature_Xenium",
  "cell_type",
  "cell_type_2",
  "pred_conf",
  "x_cell", "y_cell",
  "CHIKV_sgRNA",
  "CHIKV_5",
  "CHIKV_neg",
  "CHIKV_class"
)

xen@meta.data %>%
  as_tibble(rownames = "cell_id") %>%
  mutate(CHIKV_class = recode(cell_class, high = "CHIKV+", low = "CHIKV-")) %>%
  select(all_of(clmns)) %>%
  write_tsv(here(params$geo_dir, "xenium_metadata.tsv.gz"))
```
